// @rbx/creator-hub-search/src/clients/docSiteSearch.ts
import { useMemo } from 'react';
import type {
  SearchDocSiteRequest,
  SearchCreatorHubRequest,
  SearchCreatorHubResponse,
  SearchCreatorHubResult,
} from '@rbx/client-creator-resources-search-api/v1';
import { SearchApi } from '@rbx/client-creator-resources-search-api/v1';
import { Configuration } from '@rbx/clients-core';
import { DocumentationContentType as DocsSearchDocumentationContentType } from '@rbx/creator-docs-search';
import { useSearchConfig } from '../contexts/SearchConfigContext';
import type { TTranslateFunction } from '../localization/types/TranslateFunction';
import throwError from '../utilities/error';
import { getBEDEV2ServiceBasePath } from '../utilities/getBasePaths';
import { CategoryTitles } from './CategoryTitles';
import {
  DocumentationContentType,
  DocumentationCloudApiSubType,
  DocumentationCloudApiType,
  type DocumentationSubType,
  type DocumentationThirdType,
  type DocSiteResult,
  parseDocumentationContentType,
  parseDocumentationSubType,
  parseDocumentationThirdType,
} from './docSiteSearchType';

export type { DocSiteResult };
export { parseDocumentationContentType, parseDocumentationSubType, parseDocumentationThirdType };

/**
 * Creates a SearchApi instance with the given robloxSiteDomain.
 */
function createSearchApi(robloxSiteDomain: string): SearchApi {
  const configuration = new Configuration({
    robloxSiteDomain,
    basePath: getBEDEV2ServiceBasePath('creator-resources-search-api', robloxSiteDomain),
    credentials: 'include',
  });
  return new SearchApi(configuration);
}

export interface DocSiteSearchResponse {
  searchSessionId?: string | null;
  totalResults?: number;
  isSemantic?: boolean;
  isQueryUnderstandApplied?: boolean;
  isError: boolean;
  results: DocSiteResult[];
}

export interface SearchClient {
  search(request: SearchDocSiteRequest): Promise<DocSiteSearchResponse>;
}

function resolveCreatorHubSearchResults(response: SearchCreatorHubResponse): DocSiteSearchResponse {
  if (response.results) {
    return {
      searchSessionId: response.searchSessionId,
      totalResults: response.totalResults,
      isSemantic: true,
      isQueryUnderstandApplied: response.isQueryUnderstandApplied,
      isError: false,
      results: response.results.map((result: SearchCreatorHubResult) => ({
        elasticsearchDocumentId: result.elasticsearchDocumentId,
        identifier: result.identifier,
        title: result.title,
        displayedSummary: result.displayedSummary,
        documentationContentType: parseDocumentationContentType(result.contentType ?? ''),
        documentationSubType: parseDocumentationSubType(result.subType ?? ''),
        documentationThirdType: parseDocumentationThirdType(result.thirdType ?? ''),
        resultTargetReference: result.resultTargetReference,
        url: result.resultTargetReference ?? null,
        tags: result.tags,
        labels: result.labels,
        highlightedTitle: result.highlightedTitle,
        highlightedDisplayedSummary: result.highlightedDisplayedSummary,
        createdAtUtc: result.createdAtUtc,
        updatedAtUtc: result.updatedAtUtc,
        views: result.views,
        clicks: result.clicks,
        author: result.author ? { name: result.author.name ?? '' } : undefined,
        entityId: result.entityId ?? undefined,
        breadcrumb: result.breadcrumb ?? undefined,
        experienceName: result.entityName ?? undefined,
        creatorName: result.creatorName ?? undefined,
      })),
    };
  }
  return { isError: false, results: [] };
}

// CreatorHub results are client-side ranked so we fetch a larger batch upfront.
const CREATOR_HUB_SEARCH_LIMIT = 200;
// Defensive fallback for the backend `limit` field when building the request
// and `pageSize` is still missing (today the public API default below always
// fills it in, so this only protects future callers of `makeRequest`).
const DEFAULT_SEARCH_PAGE_SIZE = 100;
// Public API default for `SearchClient.search({ pageSize })` when the caller
// doesn't pass one — what consumers of `useDocSiteSearchClient()` actually get.
const DEFAULT_REQUEST_PAGE_SIZE = 10;

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

function hasServerErrorStatus(error: unknown): boolean {
  if (typeof error !== 'object' || error === null || !('response' in error)) {
    return false;
  }
  const resp = (error as Record<string, unknown>).response;
  if (typeof resp !== 'object' || resp === null || !('status' in resp)) {
    return false;
  }
  const status = (resp as Record<string, unknown>).status;
  return typeof status === 'number' && status >= 500 && status < 600;
}

/**
 * Creates a search client that talks to the Hub Search backend
 * (`searchSearchCreatorHub`). Results are normalized to `DocSiteResult`s
 * and re-prefixed with the request locale when applicable.
 */
export function createSearchClient(searchApi: SearchApi): SearchClient {
  return {
    async search({
      keyword,
      documentationContentType = '',
      documentationSubType = '',
      documentationThirdType = '',
      tag = '',
      isFuzzyMatch = false,
      pageSize = DEFAULT_REQUEST_PAGE_SIZE,
      pageIndex,
      locale = '',
      searchSessionId,
    }: SearchDocSiteRequest): Promise<DocSiteSearchResponse> {
      const makeRequest = async (
        requestParams: SearchDocSiteRequest,
      ): Promise<DocSiteSearchResponse> => {
        const docContentType = requestParams.documentationContentType ?? '';
        let contentType: Array<string> | undefined;
        if (docContentType === 'CreatorHub') {
          contentType = ['CreatorHub'];
        } else if (docContentType === 'Learn') {
          contentType = ['Article', 'CloudApi', 'DevForum', 'LuaApi', 'Video'];
        } else if (docContentType) {
          contentType = [docContentType];
        }

        const searchCreatorHubRequest: SearchCreatorHubRequest = {
          query: requestParams.keyword ?? '',
          locale: requestParams.locale ?? '',
          searchSessionId: requestParams.searchSessionId ?? '',
          limit:
            contentType?.length === 1 && contentType[0] === 'CreatorHub'
              ? CREATOR_HUB_SEARCH_LIMIT
              : (requestParams.pageSize ?? DEFAULT_SEARCH_PAGE_SIZE),
          isFuzzyMatch: requestParams.isFuzzyMatch ?? false,
          contentType,
          subType: requestParams.documentationSubType
            ? [requestParams.documentationSubType]
            : undefined,
          thirdType: requestParams.documentationThirdType
            ? [requestParams.documentationThirdType]
            : undefined,
          entityId: requestParams.tag ?? '',
        };
        const response = await searchApi.searchSearchCreatorHub({
          searchSearchCreatorHubRequest: searchCreatorHubRequest,
        });
        const resolved = resolveCreatorHubSearchResults(response);

        const reqLocale = (requestParams.locale ?? '').toLowerCase();
        if (reqLocale) {
          resolved.results = resolved.results.map((result) => {
            const ct = result.documentationContentType;
            if (
              ct === DocsSearchDocumentationContentType.CreatorHub ||
              ct === DocsSearchDocumentationContentType.Video ||
              ct === DocsSearchDocumentationContentType.DevForum
            ) {
              return result;
            }
            const url = result.url;
            if (!url || /^\/[a-z]{2}-[a-z0-9]{2}\//.test(url)) {
              return result;
            }
            return { ...result, url: `/${reqLocale}${url}` };
          });
        }

        return resolved;
      };

      const executeWithRetry = async (
        params: SearchDocSiteRequest,
        retryCount: number,
        hasTriedWithoutLocale: boolean,
      ): Promise<DocSiteSearchResponse> => {
        const maxRetries = 3;

        try {
          const results = await makeRequest(params);

          // If results are empty and we have a locale and haven't tried without locale yet
          if (
            (!results.results || results.results.length === 0) &&
            locale &&
            !hasTriedWithoutLocale
          ) {
            // Retry without locale
            const paramsWithoutLocale = {
              ...params,
              locale: '',
            };
            const resultsWithoutLocale = await executeWithRetry(
              paramsWithoutLocale,
              retryCount,
              true,
            );
            if (
              !resultsWithoutLocale ||
              !resultsWithoutLocale.results ||
              resultsWithoutLocale.results.length === 0
            ) {
              return {
                isError: false,
                results: [],
              };
            }

            const resultsWithLocaleResults = resultsWithoutLocale.results.map((result) => {
              const {
                elasticsearchDocumentId,
                url,
                documentationContentType: resultDocumentationContentType,
              } = result;

              if (!elasticsearchDocumentId || !url) {
                return result;
              }

              // Elasticsearch index document id format: https://github.com/Roblox/creator-docs-internal-search/blob/main/src/compile/utils/indexDocumentationUtils.ts#L16
              const elasticsearchDocumentIdSegments = elasticsearchDocumentId.split(':');
              let extractLocale =
                elasticsearchDocumentIdSegments[
                  elasticsearchDocumentIdSegments.length - 1
                ].toLowerCase();
              // Following case doesn't need to add locale prefix:
              // 1. locale is `en-us`
              // 2. url already has locale prefix
              // 3. locale is not valid
              // 4. content type is `Video`, `DevForum`, `CreatorHub`
              if (
                extractLocale === 'en-us' ||
                !/^[a-z]{2}-[a-z]{2}$/.test(extractLocale) ||
                (resultDocumentationContentType &&
                  (
                    [
                      DocsSearchDocumentationContentType.Video,
                      DocsSearchDocumentationContentType.DevForum,
                    ] as (DocumentationContentType | null)[]
                  )
                    .concat([DocumentationContentType.CreatorHub])
                    .includes(resultDocumentationContentType)) ||
                /^\/[a-z]{2}-[a-z]{2}\/.*$/.test(url)
              ) {
                extractLocale = '';
              }

              return {
                ...result,
                url: extractLocale ? `/${extractLocale}${url}` : url,
              };
            });

            return {
              ...resultsWithoutLocale,
              results: resultsWithLocaleResults,
            };
          }

          return results;
        } catch (error) {
          const isServerError = hasServerErrorStatus(error);

          if (isServerError && retryCount < maxRetries) {
            await delay(200 * (retryCount + 1));
            return executeWithRetry(params, retryCount + 1, hasTriedWithoutLocale);
          }

          throwError(
            `searchClient: failed to fetch search results ${error instanceof Error ? error.message : String(error)}`,
          );
          return {
            isError: true,
            results: [],
          };
        }
      };

      const searchDocSiteRequest = {
        keyword,
        isFuzzyMatch,
        documentationContentType,
        documentationSubType,
        documentationThirdType,
        tag,
        pageIndex,
        pageSize,
        locale,
        searchSessionId,
      } as SearchDocSiteRequest;

      return executeWithRetry(searchDocSiteRequest, 0, false);
    },
  };
}

/**
 * Hook that returns a search client configured with the current environment.
 * Uses SearchConfigContext to get the robloxSiteDomain.
 *
 * @example
 * ```tsx
 * const searchClient = useDocSiteSearchClient();
 * const results = await searchClient.search({ keyword: "test" });
 * ```
 */
export function useDocSiteSearchClient(): SearchClient {
  const { robloxSiteDomain } = useSearchConfig();

  return useMemo(() => createSearchClient(createSearchApi(robloxSiteDomain)), [robloxSiteDomain]);
}

export function getCategoryName(result: DocSiteResult): string | null {
  const {
    documentationContentType: contentType,
    documentationSubType: subType,
    documentationThirdType: thirdType,
  } = result;
  if (contentType === null) {
    return null;
  }
  if (contentType === DocumentationContentType.LuaAPI) {
    const engineTitle = CategoryTitles.Engine;
    if (subType === null) {
      return engineTitle;
    }
    if (thirdType === null) {
      return `${engineTitle}: ${subType}`;
    }
    return `${engineTitle}: ${subType}: ${thirdType}`;
  }
  if (contentType === DocumentationContentType.Article) {
    return subType;
  }
  if (contentType === DocumentationContentType.CloudAPI) {
    return CategoryTitles.Cloud;
  }
  if (contentType === DocumentationContentType.Video) {
    return CategoryTitles.Video;
  }
  if (contentType === DocumentationContentType.DevForum) {
    const devForumTitle = CategoryTitles.DevForum;
    if (subType === null) {
      return devForumTitle;
    }
    if (thirdType === null) {
      return `${devForumTitle}: ${subType}`;
    }
    return `${devForumTitle}: ${subType}: ${thirdType}`;
  }
  return null;
}

export function getCategoryTranslationLabels(result: DocSiteResult): string[] {
  const {
    documentationContentType: contentType,
    documentationSubType: subType,
    documentationThirdType: thirdType,
  } = result;
  if (contentType === null) {
    return [];
  }
  if (contentType === DocumentationContentType.CreatorHub) {
    return ['Label.CreatorHub'];
  }
  if (contentType === DocumentationContentType.CloudAPI) {
    const labels = ['Label.Cloud'];
    const cloudApiSubTypeLabels: Partial<Record<DocumentationSubType, string>> = {
      [DocumentationCloudApiType.Legacy]: 'Label.CloudAPIVersionLegacy',
      [DocumentationCloudApiType.V1]: 'Label.CloudAPIVersionV1',
      [DocumentationCloudApiType.V2]: 'Label.CloudAPIVersionV2',
      [DocumentationCloudApiType.Features]: 'Label.CloudAPIVersionFeatures',
    };
    if (subType === null || !cloudApiSubTypeLabels[subType]) {
      return labels;
    }
    labels.push(cloudApiSubTypeLabels[subType]);
    const cloudApiThirdTypeLabels: Partial<Record<DocumentationThirdType, string>> = {
      [DocumentationCloudApiSubType.Endpoint]: 'Label.CloudAPITypeEndpoint',
      [DocumentationCloudApiSubType.API]: 'Label.CloudAPITypeAPI',
      [DocumentationCloudApiSubType.Resource]: 'Label.CloudAPITypeResource',
    };
    if (thirdType !== null && cloudApiThirdTypeLabels[thirdType]) {
      labels.push(cloudApiThirdTypeLabels[thirdType]);
    }
    return labels;
  }
  return [];
}

export const getTranslatedCategoryDisplayText = (
  result: DocSiteResult,
  translate: TTranslateFunction,
): string | null => {
  const {
    documentationContentType: contentType,
    documentationSubType: subType,
    documentationThirdType: thirdType,
  } = result;
  if (contentType === null) {
    return null;
  }
  if (contentType === DocumentationContentType.LuaAPI) {
    if (subType === null) {
      return translate(`Label.Engine`);
    }
    if (thirdType === null) {
      return translate(`Label.${subType}`);
    }
    return `${translate(`Label.${subType}`)} ${translate(`Label.${thirdType}`)}`;
  }
  if (contentType === DocumentationContentType.Article) {
    return translate(`Label.${subType}`);
  }
  if (contentType === DocumentationContentType.CloudAPI) {
    return getCategoryTranslationLabels(result)
      .map((label) => translate(label))
      .join(' ');
  }
  if (contentType === DocumentationContentType.Video) {
    return translate('Label.Video');
  }
  if (contentType === DocumentationContentType.DevForum) {
    const devForumLabel = translate(`Label.DevForum`);
    if (subType === null) {
      return devForumLabel;
    }
    if (thirdType === null) {
      return `${devForumLabel} ${translate(`Label.${subType}`)}`;
    }
    return `${devForumLabel} ${translate(`Label.${subType}`)} ${translate(`Label.${thirdType}`)}`;
  }
  return `${contentType} ${subType ?? ''} ${thirdType ?? ''}`;
};
