// @rbx/creator-hub-search/src/clients/toolboxSearchClient.ts
import { useMemo } from 'react';
import type {
  CreatorStoreAsset,
  SearchCreatorStoreAssetsResponse,
  SearchCreatorStoreAssetsRequest,
} from '@rbx/client-toolbox-service/v1';
import { SearchCategoryType, ToolboxApi } from '@rbx/client-toolbox-service/v1';
import { Configuration } from '@rbx/clients-core';
import { useSearchConfig } from '../contexts/SearchConfigContext';
import throwError from '../utilities/error';
import { getBEDEV2ServiceBasePath, getCreatorHubBaseUrl } from '../utilities/getBasePaths';

/**
 * Client for the Toolbox Search Creator Store Assets API:
 * https://create.roblox.com/docs/cloud/reference/features/creator-store#Toolbox_SearchCreatorStoreAssets
 */

export interface ToolboxPrice {
  /** ISO 4217 currency code (e.g. `'USD'`, `'EUR'`) or `'Robux'` for Robux. */
  currencyCode: string;
  /**
   * Decimal-aware amount. For Robux this is an integer (e.g. `10`); for fiat
   * currencies the minor units are folded in via `significand * 10^exponent`
   * (e.g. `significand: 199, exponent: -2` → `1.99`).
   */
  amount: number;
}

export interface ToolboxSearchResult {
  id: number;
  name: string;
  description: string;
  assetTypeId: number | null;
  categoryType: SearchCategoryType;
  categoryPath: string | null;
  /**
   * Zero-based index of the source category in TOOLBOX_SEARCH_CATEGORIES. Drives
   * the 2-per-category ordering used by the Store "all results" view.
   */
  categoryIndex?: number;
  creatorName: string;
  /** First image preview asset id, when available; UI can hydrate via thumbnails service. */
  thumbnailAssetId: number | null;
  /** Creator Hub Store detail URL, e.g. https://create.roblox.com/store/asset/123. */
  url: string;
  /**
   * `true` when the asset is free to acquire — either no purchase price was
   * returned, or the price quantity is zero. Consumers should render a
   * "Free" label in this case (price will be `null`).
   */
  isFree: boolean;
  /**
   * Currency + amount when the API surfaced a non-zero purchase price.
   * `null` when the asset is free or when the price payload couldn't be
   * parsed into a numeric quantity.
   */
  price: ToolboxPrice | null;
}

export interface ToolboxSearchResponse {
  isError: boolean;
  results: ToolboxSearchResult[];
}

export interface ToolboxSearchRequest {
  query: string;
  /** Total combined cap across categories. Defaults to 10. */
  limit?: number;
}

export interface ToolboxSearchClient {
  search(request: ToolboxSearchRequest): Promise<ToolboxSearchResponse>;
}

const DEFAULT_LIMIT = 10;

interface ToolboxSearchCategoryEntry {
  categoryType: SearchCategoryType;
  categoryPath: string | null;
}

// Categories we surface in the search dialog. Order matters for round-robin
// interleaving (each category gets a slot in turn until the limit is hit).
const TOOLBOX_SEARCH_CATEGORIES: ToolboxSearchCategoryEntry[] = [
  { categoryType: SearchCategoryType.Model, categoryPath: null },
  { categoryType: SearchCategoryType.Model, categoryPath: '3d' },
  { categoryType: SearchCategoryType.Audio, categoryPath: null },
  { categoryType: SearchCategoryType.Plugin, categoryPath: null },
  { categoryType: SearchCategoryType.Model, categoryPath: 'gameplay' },
  { categoryType: SearchCategoryType.Decal, categoryPath: '2d' },
  { categoryType: SearchCategoryType.Model, categoryPath: 'visual-effect' },
];

const MAX_RETRIES = 3;
const BASE_RETRY_DELAY_MS = 200;

const delay = (ms: number): Promise<void> =>
  new Promise((resolve) => {
    setTimeout(resolve, ms);
  });

function hasServerErrorStatus(error: unknown): boolean {
  if (typeof error !== 'object' || error === null || !('response' in error)) {
    return false;
  }
  const resp = (error as Record<string, unknown>).response;
  if (typeof resp !== 'object' || resp === null || !('status' in resp)) {
    return false;
  }
  const status = (resp as Record<string, unknown>).status;
  return typeof status === 'number' && status >= 500 && status < 600;
}

/**
 * Creates a ToolboxApi instance pointed at the given robloxSiteDomain.
 */
function createToolboxApi(robloxSiteDomain: string): ToolboxApi {
  const configuration = new Configuration({
    robloxSiteDomain,
    basePath: getBEDEV2ServiceBasePath('toolbox-service', robloxSiteDomain),
    credentials: 'include',
  });
  return new ToolboxApi(configuration);
}

/**
 * Resolves the currency code + decimal amount from a Money payload.
 *
 * Returns `null` when the payload is missing or carries no numeric quantity.
 * `currencyCode` falls back to `'Robux'` when the API omits it (some free
 * Robux listings come back without an explicit code).
 *
 * `amount` is computed as `significand * 10^exponent`:
 *   - Robux prices typically have `exponent === 0` → integer amount.
 *   - Fiat prices typically have `exponent === -2` → decimal amount
 *     (e.g. `significand: 199, exponent: -2` → `1.99`).
 */
/** @internal Exported for unit testing only. */
export function extractPrice(
  purchasePrice:
    | {
        currencyCode?: string | null;
        quantity?: { significand?: number; exponent?: number } | null;
      }
    | null
    | undefined,
): ToolboxPrice | null {
  if (!purchasePrice) {
    return null;
  }
  const significand = purchasePrice.quantity?.significand;
  const exponent = purchasePrice.quantity?.exponent ?? 0;
  if (typeof significand !== 'number') {
    return null;
  }
  const rawAmount = significand * 10 ** exponent;
  if (!Number.isFinite(rawAmount) || rawAmount < 0) {
    return null;
  }
  const currencyCode = purchasePrice.currencyCode ?? 'Robux';
  // Round Robux to integer; preserve fiat decimals up to the exponent.
  const amount =
    currencyCode === 'Robux'
      ? Math.round(rawAmount)
      : Number(rawAmount.toFixed(Math.max(0, -Math.min(exponent, 0))));
  return { currencyCode, amount };
}

/**
 * Maps a raw CreatorStoreAsset into the trimmed shape consumed by the search UI.
 */
/** @internal Exported for unit testing only. */
export function mapAssetToResult(
  asset: CreatorStoreAsset,
  categoryType: SearchCategoryType,
  categoryPath: string | null,
  creatorHubBaseUrl: string,
): ToolboxSearchResult | null {
  const inner = asset.asset;
  if (!inner) {
    return null;
  }
  // The discriminated union narrows by asset type, but every variant carries the base `id` / `name` fields.
  const innerRecord = inner as {
    id?: number;
    name?: string | null;
    description?: string | null;
    assetTypeId?: number | null;
    previewAssets?: { imagePreviewAssets?: number[] | null } | null;
  };
  const id = innerRecord.id;
  if (typeof id !== 'number') {
    return null;
  }
  const thumbnailAssetId = innerRecord.previewAssets?.imagePreviewAssets?.[0] ?? null;
  const product = asset.creatorStoreProduct;
  const parsedPrice = extractPrice(product?.purchasePrice);
  // Classification:
  //   - No purchasePrice payload at all          → Free (price = null).
  //   - Payload with amount === 0                → Free (price = null).
  //   - Payload with amount > 0 (Robux or fiat)  → paid; surface currency + amount.
  //   - Payload exists but unparseable           → treat as Free so we don't
  //     leave the row with no second-line context.
  let isFree: boolean;
  let price: ToolboxPrice | null;
  if (!product?.purchasePrice) {
    isFree = true;
    price = null;
  } else if (!parsedPrice || parsedPrice.amount === 0) {
    isFree = true;
    price = null;
  } else {
    isFree = false;
    price = parsedPrice;
  }
  return {
    id,
    name: innerRecord.name ?? '',
    description: innerRecord.description ?? '',
    assetTypeId: innerRecord.assetTypeId ?? null,
    categoryType,
    categoryPath,
    creatorName: asset.creator?.name ?? '',
    thumbnailAssetId,
    url: `${creatorHubBaseUrl}/store/asset/${id}`,
    isFree,
    price,
  };
}

/**
 * Round-robin interleave so each category is represented in the merged list.
 */
/** @internal Exported for unit testing only. */
export function interleaveResults(
  resultsByCategory: ToolboxSearchResult[][],
  limit: number,
): ToolboxSearchResult[] {
  const merged: ToolboxSearchResult[] = [];
  const seen = new Set<number>();
  const cursors = resultsByCategory.map(() => 0);
  let done = false;
  while (!done && merged.length < limit) {
    done = true;
    for (let i = 0; i < resultsByCategory.length; i += 1) {
      const list = resultsByCategory[i];
      while (cursors[i] < list.length && seen.has(list[cursors[i]].id)) {
        cursors[i] += 1;
      }
      const cursor = cursors[i];
      if (cursor < list.length) {
        seen.add(list[cursor].id);
        merged.push(list[cursor]);
        cursors[i] = cursor + 1;
        done = false;
        if (merged.length >= limit) {
          break;
        }
      }
    }
  }
  return merged;
}

/**
 * Picks a single owner category per asset id, preferring categories that carry a
 * `categoryPath` (3D / Gameplay / VFX / 2D) over the general null-path category,
 * then returns each category's list filtered to the assets it owns.
 *
 * The general "Models" query (no categoryPath) is a superset of the path-specific
 * Model queries, so without this an overlapping asset would be kept under
 * "Models" (visited first) and stripped of its subcategory label. Preferring the
 * path-specific category keeps e.g. a 3D model labeled "3D Assets" and grouped
 * under "3D"; "Models" then surfaces only assets without a curated subcategory.
 * Ties (both path-specific or both null-path) keep the lower-index category.
 */
/** @internal Exported for unit testing only. */
export function dedupeByPreferredCategory(
  resultsByCategory: ToolboxSearchResult[][],
  categoryHasPath: boolean[],
): ToolboxSearchResult[][] {
  const ownerIndexById = new Map<number, number>();
  resultsByCategory.forEach((list, index) => {
    list.forEach((result) => {
      const currentOwner = ownerIndexById.get(result.id);
      if (currentOwner === undefined) {
        ownerIndexById.set(result.id, index);
      } else if (categoryHasPath[index] && !categoryHasPath[currentOwner]) {
        ownerIndexById.set(result.id, index);
      }
    });
  });
  return resultsByCategory.map((list, index) =>
    list.filter((result) => ownerIndexById.get(result.id) === index),
  );
}

interface CreateToolboxSearchClientConfig {
  robloxSiteDomain: string;
}

/**
 * Creates a ToolboxSearchClient using the supplied ToolboxApi.
 */
export function createToolboxSearchClient(
  toolboxApi: ToolboxApi,
  { robloxSiteDomain }: CreateToolboxSearchClientConfig,
): ToolboxSearchClient {
  const creatorHubBaseUrl = getCreatorHubBaseUrl(robloxSiteDomain);

  const searchOneCategory = async (
    query: string,
    categoryType: SearchCategoryType,
    categoryPath: string | null,
    categoryIndex: number,
    maxPageSize: number,
  ): Promise<ToolboxSearchResult[]> => {
    const body: SearchCreatorStoreAssetsRequest = {
      searchCategoryType: categoryType,
      query,
      maxPageSize,
      ...(categoryPath != null && { categoryPath }),
    };

    const executeWithRetry = async (
      retryCount: number,
    ): Promise<SearchCreatorStoreAssetsResponse> => {
      try {
        return await toolboxApi.toolboxSearchCreatorStoreAssets({
          searchCreatorStoreAssetsRequest: body,
        });
      } catch (error) {
        if (hasServerErrorStatus(error) && retryCount < MAX_RETRIES) {
          await delay(BASE_RETRY_DELAY_MS * (retryCount + 1));
          return executeWithRetry(retryCount + 1);
        }
        throw error;
      }
    };

    const response = await executeWithRetry(0);
    const assets = response.creatorStoreAssets ?? [];
    return assets
      .map((asset) => mapAssetToResult(asset, categoryType, categoryPath, creatorHubBaseUrl))
      .filter((r): r is ToolboxSearchResult => r !== null)
      .map((result) => ({ ...result, categoryIndex }));
  };

  return {
    async search({
      query,
      limit = DEFAULT_LIMIT,
    }: ToolboxSearchRequest): Promise<ToolboxSearchResponse> {
      const trimmed = query.trim();
      if (trimmed.length === 0) {
        return { isError: false, results: [] };
      }

      // Give each category an equal share of the cap (ceil so no category is
      // starved when limit < category count). interleaveResults then round-robins
      // and dedups down to `limit`, biasing earlier categories on overflow.
      const perCategoryLimit = Math.max(1, Math.ceil(limit / TOOLBOX_SEARCH_CATEGORIES.length));

      const settled = await Promise.allSettled(
        TOOLBOX_SEARCH_CATEGORIES.map(({ categoryType, categoryPath }, index) =>
          searchOneCategory(trimmed, categoryType, categoryPath, index, perCategoryLimit),
        ),
      );

      const failureMessages: string[] = [];
      const resultsByCategory = settled.map((result) => {
        if (result.status === 'fulfilled') {
          return result.value;
        }
        failureMessages.push(
          result.reason instanceof Error ? result.reason.message : String(result.reason),
        );
        return [];
      });

      // Give each overlapping asset to its most specific category before merging,
      // so 3D/Gameplay/VFX models keep their subcategory label instead of being
      // absorbed by the general (null-path) "Models" category.
      const ownedResultsByCategory = dedupeByPreferredCategory(
        resultsByCategory,
        TOOLBOX_SEARCH_CATEGORIES.map((category) => category.categoryPath != null),
      );

      const merged = interleaveResults(ownedResultsByCategory, limit);
      // Treat partial failures as non-error if we still got results; only flag
      // (and report) an error when every category failed. Reporting once here —
      // rather than inside the map above — avoids per-category Sentry spam on
      // partial failures and keeps the dev-only throw from escaping mid-map and
      // discarding the categories that succeeded.
      const isError = failureMessages.length > 0 && merged.length === 0;
      if (isError) {
        throwError(
          `toolboxSearchClient: failed to fetch creator store assets: ${failureMessages.join('; ')}`,
        );
      }
      return { isError, results: merged };
    },
  };
}

/**
 * Hook that returns a ToolboxSearchClient configured for the current environment.
 *
 * @example
 * ```tsx
 * const toolboxSearchClient = useToolboxSearchClient();
 * const { results } = await toolboxSearchClient.search({ query: 'car' });
 * ```
 */
export function useToolboxSearchClient(): ToolboxSearchClient {
  const { robloxSiteDomain } = useSearchConfig();
  return useMemo(
    () => createToolboxSearchClient(createToolboxApi(robloxSiteDomain), { robloxSiteDomain }),
    [robloxSiteDomain],
  );
}
