// @rbx/creator-hub-navigation/src/topNavigation/components/authenticationStatus/AuthenticationStatusMenu.tsx
import type { Dispatch, FunctionComponent, ReactElement, SetStateAction } from 'react';
import React, { useCallback } from 'react';
import { useRobloxAuthentication } from '@rbx/auth';
import type { TMenuItemProps } from '@rbx/foundation-ui';
import { Menu, MenuLabel, MenuSection, MenuSeparator } from '@rbx/foundation-ui';
import { useTranslation } from '@rbx/intl';
import type { AccountSwitcherSnackbarConfig } from '../../../accountSwitcher';
import { isRobloxApp } from '../../../accountSwitcher/utils/browser';
import {
  clickNavCopyUserIdEventModel,
  clickNavLogOutEventModel,
  clickNavSettingsEventModel,
  clickNavSwitchAccountsEventModel,
} from '../../../event/eventConstants';
import useNavigationConfigs from '../../../hooks/useNavigationConfigs';
import type { TBuildTarget, TProductKey, TRobloxEnvironment } from '../../../types';
import { getCreatorHubBasePathV2 as getCreatorHubBasePath } from '../../../utils/getBasePaths';
import isDashboard from '../../../utils/isDashboard';
import AuthenticationStatusMenuItem from './AuthenticationStatusMenuItem';

// we can't use useProductUrls because devforum doesn't support NextJS routing
const getSettingsHref = (
  currentProduct: TProductKey,
  target: TBuildTarget,
  robloxEnvironment: TRobloxEnvironment,
) => {
  const baseUrl = getCreatorHubBasePath(target, robloxEnvironment);
  const creatorHubBasePath = isDashboard(currentProduct) ? '/' : baseUrl;
  return `${creatorHubBasePath}settings/preferences`;
};

type TAuthenticationStatusMenuProps = {
  desktopDropdownContent?: ReactElement<TMenuItemProps>[];
  isAccountSwitcherFrameLoaded: boolean; // whether the account switcher iframe has loaded. if false, we show a disabled menu item with a loading state. if true, we show the menu item based on whether the account switcher is enabled.
  isAccountSwitcherEnabled: boolean; // whether the account switcher is enabled. This comes from the iframe itself.
  isAccountSwitcherSupported: boolean; // whether the account switcher is supported in the current environment. Determines whether to show the account switcher menu item and whether to load the account switcher iframe.
  onLogout?: VoidFunction; // optional callback that runs after logout
  setIsDialogOpen: (isOpen: boolean) => void;
  setIsLogoutInProgress: Dispatch<SetStateAction<boolean>>;
  setIsPopoverOpen: Dispatch<SetStateAction<boolean>>;
  setSnackbarConfig: Dispatch<SetStateAction<AccountSwitcherSnackbarConfig | undefined>>;
};

const AuthenticationStatusMenu: FunctionComponent<TAuthenticationStatusMenuProps> = ({
  desktopDropdownContent,
  isAccountSwitcherFrameLoaded,
  isAccountSwitcherEnabled,
  isAccountSwitcherSupported,
  onLogout,
  setIsDialogOpen,
  setIsLogoutInProgress,
  setIsPopoverOpen,
  setSnackbarConfig,
}) => {
  const { translate } = useTranslation();
  const { currentProduct, target, robloxEnvironment, sendEvent } = useNavigationConfigs();
  const { logout, user } = useRobloxAuthentication();

  const settingsHref = getSettingsHref(currentProduct, target, robloxEnvironment);

  const onClickSettings = useCallback(() => {
    sendEvent(clickNavSettingsEventModel);
  }, [sendEvent]);

  const onClickCopyUserId = useCallback(() => {
    if (user?.id) {
      sendEvent(clickNavCopyUserIdEventModel);
      void navigator.clipboard.writeText(user.id.toString());
    }
    setSnackbarConfig({ title: translate('Message.UserIdCopied'), isError: false });
    setIsPopoverOpen(false); // there's a bug where the popover doesn't close if there's no href, so we have to close it manually here
  }, [sendEvent, setIsPopoverOpen, setSnackbarConfig, translate, user?.id]);

  const onClickSwitchAccounts = useCallback(() => {
    sendEvent(clickNavSwitchAccountsEventModel);
    setIsDialogOpen(true);
  }, [sendEvent, setIsDialogOpen]);

  const onClickLogout = useCallback(async () => {
    sendEvent(clickNavLogOutEventModel);
    setIsLogoutInProgress(true);
    try {
      await logout();
    } catch {
      // fail silently
    }
    if (onLogout) {
      onLogout();
    }
    setIsLogoutInProgress(false);
  }, [logout, onLogout, sendEvent, setIsLogoutInProgress]);

  const shouldShowLogOut = !isRobloxApp();

  return (
    <Menu size='Medium' className='min-width-[260px]'>
      <MenuSection>
        <MenuLabel title={translate('Label.SignedInAs', { username: user?.name ?? '' })} disabled />
        <AuthenticationStatusMenuItem
          text={translate('Heading.Settings')}
          onSelect={onClickSettings}
          href={settingsHref}
        />
      </MenuSection>
      <MenuSeparator />
      {desktopDropdownContent && desktopDropdownContent.length > 0 && (
        <>
          <MenuSection>{desktopDropdownContent}</MenuSection>
          <MenuSeparator />
        </>
      )}
      <MenuSection>
        <AuthenticationStatusMenuItem
          key='CopyUserId'
          text={translate('Action.CopyUserId')}
          onSelect={onClickCopyUserId}
        />
        {isAccountSwitcherSupported &&
        (!isAccountSwitcherFrameLoaded || isAccountSwitcherEnabled) ? ( // if frame is loaded, show only if enabled. otherwise show disabled menu item while loading.
          <AuthenticationStatusMenuItem
            key='SwitchAccounts'
            text={translate('Action.SwitchAccounts')}
            onSelect={onClickSwitchAccounts}
            disabled={!isAccountSwitcherEnabled}
          />
        ) : null}
      </MenuSection>
      {shouldShowLogOut && (
        <>
          <MenuSeparator />
          <MenuSection>
            <AuthenticationStatusMenuItem
              key='LogOut'
              text={translate('Action.LogOut')}
              onSelect={onClickLogout}
            />
          </MenuSection>
        </>
      )}
    </Menu>
  );
};

export default AuthenticationStatusMenu;
