// @rbx/google-gax@0.1.3/build/src/auth/authclient.js
"use strict";
// Copyright 2012 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright 2023 Roblox Corporation
//
// This file was derived and modified from src/auth/authclient.ts
// in github.com/googleapis/google-auth-library-nodejs.
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuthClient = exports.DEFAULT_EAGER_REFRESH_THRESHOLD_MILLIS = void 0;
const events_1 = require("events");
const transporters_1 = require("./transporters");
/**
 * The default {@link AuthClientOptions.eagerRefreshThresholdMillis}
 */
exports.DEFAULT_EAGER_REFRESH_THRESHOLD_MILLIS = 0;
class AuthClient extends events_1.EventEmitter {
    constructor(opts = {}) {
        var _a, _b;
        super();
        this._credentials = {};
        this.eagerRefreshThresholdMillis = exports.DEFAULT_EAGER_REFRESH_THRESHOLD_MILLIS;
        this.forceRefreshOnFailure = false;
        // Shared auth options
        this._credentials = {};
        // Shared client options
        this.transporter = (_a = opts.transporter) !== null && _a !== void 0 ? _a : new transporters_1.DefaultTransporter();
        if (opts.transporterOptions) {
            this.transporter.defaults = opts.transporterOptions;
        }
        if (opts.eagerRefreshThresholdMillis) {
            this.eagerRefreshThresholdMillis = opts.eagerRefreshThresholdMillis;
        }
        this.forceRefreshOnFailure = (_b = opts.forceRefreshOnFailure) !== null && _b !== void 0 ? _b : false;
    }
    /**
     * Gets the auth credentials.
     */
    get credentials() {
        return Object.assign({}, this._credentials);
    }
    /**
     * Sets the auth credentials.
     */
    set credentials(credentials) {
        this._credentials = Object.assign({}, credentials);
    }
}
exports.AuthClient = AuthClient;
