// @rbx/not-approved-page-events@0.1.0/dist/index.js
// src/index.ts
var EventTypes = /* @__PURE__ */ ((EventTypes2) => {
  EventTypes2["AccountReactivationPageRendered"] = "accountReactivationPageRendered";
  EventTypes2["PageRendered"] = "pageRendered";
  EventTypes2["UnmappedViolationKey"] = "unmappedViolationKey";
  EventTypes2["CheckboxChecked"] = "checkboxChecked";
  EventTypes2["LogoutClicked"] = "logoutClicked";
  EventTypes2["ReactivateClicked"] = "reactivateClicked";
  EventTypes2["ParentVerificationClicked"] = "parentVerificationClicked";
  EventTypes2["EmailVerificationClicked"] = "emailVerificationClicked";
  EventTypes2["TermsOfUseClicked"] = "termsOfUseClicked";
  EventTypes2["AppealsProcessClicked"] = "appealsProcessClicked";
  EventTypes2["CommunityGuidelineClicked"] = "communityGuidelineClicked";
  EventTypes2["UGCGuidelinesClicked"] = "ugcGuidelinesClicked";
  EventTypes2["AppealsPortalClicked"] = "appealsPortalClicked";
  EventTypes2["ContinueClicked"] = "continueClicked";
  EventTypes2["BackClicked"] = "backClicked";
  EventTypes2["SecondChanceReactivateClicked"] = "secondChanceReactivateClicked";
  EventTypes2["MissingTranslation"] = "missingTranslation";
  EventTypes2["Error"] = "error";
  EventTypes2["VerificationRedirectRendered"] = "verificationRedirectRendered";
  EventTypes2["AccountReactivationRedirectRendered"] = "accountReactivationRedirectRendered";
  return EventTypes2;
})(EventTypes || {});
var EventContext = /* @__PURE__ */ ((EventContext2) => {
  EventContext2["NotApprovedPage"] = "NotApprovedPageV2";
  return EventContext2;
})(EventContext || {});
var EVENT_NAME = "NotApprovedPageEvent";
export {
  EVENT_NAME,
  EventContext,
  EventTypes
};
