// @rbx/cueing/core/analytics.ts
import type { CueModalId } from './ports';
import {
  type CueConfig,
  CueHiveEventType,
  CueLifecycle,
  type CueLifecycleEvent,
  type HiveEventType,
} from './types';

const lifecycleToHiveEventType = (lifecycle: CueLifecycleEvent): HiveEventType => {
  if (lifecycle === CueLifecycle.Shown) {
    return CueHiveEventType.Impression;
  }
  if (lifecycle === CueLifecycle.Finished) {
    return CueHiveEventType.Complete;
  }
  return CueHiveEventType.Dismissal;
};

const buildCueAnalyticsParameters = ({
  analyticsDefaults,
  analyticsOverrides,
  lifecycle,
  modalId,
  stepKey,
}: {
  analyticsDefaults?: Record<string, string>;
  analyticsOverrides?: Record<string, string>;
  lifecycle: CueLifecycleEvent;
  modalId: CueModalId;
  stepKey?: string;
}): Record<string, string> => {
  const eventType = lifecycleToHiveEventType(lifecycle);
  return {
    ...analyticsDefaults,
    ...analyticsOverrides,
    eventType,
    modalId,
    ...(stepKey ? { stepKey } : {}),
  };
};

export const logCueAnalytics = ({
  analyticsOverrides,
  config,
  lifecycle,
  modalId,
  stepKey,
}: {
  config: CueConfig;
  lifecycle: CueLifecycleEvent;
  modalId: CueModalId;
  stepKey?: string;
  analyticsOverrides?: Record<string, string>;
}): Record<string, string> =>
  buildCueAnalyticsParameters({
    analyticsDefaults: config.analyticsDefaults,
    analyticsOverrides,
    lifecycle,
    modalId,
    stepKey,
  });
