// @rbx/cueing/core/configure.ts
import type { CueingConfiguration } from './ports';

let configuration: CueingConfiguration | null = null;
const listeners = new Set<() => void>();

const notifyConfigurationListeners = (): void => {
  listeners.forEach((listener) => {
    listener();
  });
};

export const subscribeCueingConfiguration = (listener: () => void): (() => void) => {
  listeners.add(listener);
  return () => {
    listeners.delete(listener);
  };
};

export const getCueingConfigurationSnapshot = (): CueingConfiguration | null => configuration;

export const configureCueing = (config: CueingConfiguration): void => {
  configuration = config;
  notifyConfigurationListeners();
};

export const tryGetCueingConfiguration = (): CueingConfiguration | null => configuration;

export const getCueingConfiguration = (): CueingConfiguration => {
  if (!configuration) {
    throw new Error('Cueing is not configured. Call configureCueing before using cue APIs.');
  }
  return configuration;
};

export const resetCueingConfiguration = (): void => {
  configuration = null;
  notifyConfigurationListeners();
};
