// @rbx/cueing/core/emitCue.ts
import { logCueAnalytics } from './analytics';
import { tryGetCueingConfiguration } from './configure';
import {
  clearModalPersistDeduped,
  hasDedupedModalPersist,
  markModalPersistDeduped,
} from './persistDedupe';
import type { CueModalId } from './ports';
import type { CueLifecycleEvent } from './types';

export type EmitCueParams = {
  analyticsOverrides?: Record<string, string>;
  lifecycle: CueLifecycleEvent;
  modalId: CueModalId;
  stepKey?: string;
};

const shouldLogAnalytics = (
  lifecycle: EmitCueParams['lifecycle'],
  emitAnalyticsOn: readonly EmitCueParams['lifecycle'][] | undefined,
): boolean => {
  if (!emitAnalyticsOn) {
    return true;
  }
  return emitAnalyticsOn.includes(lifecycle);
};

const shouldPersistLifecycle = (
  lifecycle: EmitCueParams['lifecycle'],
  persistSeenOn: readonly EmitCueParams['lifecycle'][],
): boolean => persistSeenOn.includes(lifecycle);

export const emitCue = async ({
  analyticsOverrides,
  lifecycle,
  modalId,
  stepKey,
}: EmitCueParams): Promise<void> => {
  const configuration = tryGetCueingConfiguration();
  if (!configuration) {
    return;
  }

  const { analytics, modalHistory, onError, registry } = configuration;
  const config = registry[modalId];
  if (!config) {
    onError.log({ context: 'emitCue', modalId, reason: 'unknown_modal_id' });
    return;
  }

  if (stepKey && config.stepKeys && !config.stepKeys.includes(stepKey)) {
    onError.log({
      context: 'emitCue',
      modalId,
      reason: 'unknown_step_key',
      stepKey,
    });
    return;
  }

  if (shouldLogAnalytics(lifecycle, config.emitAnalyticsOn)) {
    try {
      const parameters = logCueAnalytics({
        analyticsOverrides,
        config,
        lifecycle,
        modalId,
        stepKey,
      });
      analytics.logLifecycle(lifecycle, parameters);
    } catch (error) {
      onError.log({ context: 'emitCue_analytics', lifecycle, modalId }, error);
    }
  }

  if (!shouldPersistLifecycle(lifecycle, config.persistSeenOn)) {
    return;
  }

  if (hasDedupedModalPersist(modalId)) {
    return;
  }

  markModalPersistDeduped(modalId);

  try {
    await modalHistory.recordUserSeenModal(modalId);
  } catch (error) {
    clearModalPersistDeduped(modalId);
    onError.log({ context: 'emitCue_persist', lifecycle, modalId }, error);
  }
};
