// @rbx/creator-hub-navigation/src/queries/useGetGroupsList.ts
import { useMemo } from 'react';
import { useQuery } from '@tanstack/react-query';
import { useRobloxAuthentication as useAuthentication } from '@rbx/auth';
import { GroupsApi, GroupListSurface } from '@rbx/client-creator-home-api/v1';
import { Configuration } from '@rbx/clients-core';
import useNavigationConfigs from '../hooks/useNavigationConfigs';
import { getBEDEV2ServiceBasePathV2 } from '../utils/getBasePaths';

export const getGroupsQueryKey = ['creatorHome', 'groupList', GroupListSurface.CreatorHub];

const useGetGroupsList = () => {
  const { target, robloxEnvironment: environment } = useNavigationConfigs();
  const { user } = useAuthentication();
  const groupsApi = useMemo(() => {
    const basePath = getBEDEV2ServiceBasePathV2('creator-home-api', target, environment);

    const configuration = new Configuration({
      basePath,
      credentials: 'include',
    });

    return new GroupsApi(configuration);
  }, [environment, target]);

  return useQuery({
    queryKey: getGroupsQueryKey,
    enabled: !!user,
    queryFn: () => {
      return groupsApi.groupsListGroups({
        surface: GroupListSurface.CreatorHub,
      });
    },
  });
};

export default useGetGroupsList;
