// @rbx/intl/src/components/Translate.tsx
import type { ComponentType, ReactNode } from 'react';
import React, { useContext } from 'react';
import useTranslation from '../hooks/useTranslation';
import translationResourceProviderContext from '../TranslationResourceProviderContext';
import withTranslation from './withTranslation';

interface TranslateInnerProps {
  namespace: string;
  translationKey: string;
  args?: { [key: string]: string };
  fallback?: ReactNode;
}

const TranslateInner = ({
  namespace,
  translationKey,
  args,
  fallback = null,
}: TranslateInnerProps) => {
  const { translateWithNamespace, ready } = useTranslation();

  if (!ready) {
    return fallback;
  }

  const result = translateWithNamespace(namespace, translationKey, args);
  return result || fallback;
};

const wrappedComponentCache = new Map<string, ComponentType<TranslateInnerProps>>();

function getOrCreateWrapped(namespace: string): ComponentType<TranslateInnerProps> {
  let wrapped = wrappedComponentCache.get(namespace);
  if (!wrapped) {
    wrapped = withTranslation(TranslateInner, [namespace]);
    wrapped.displayName = `Translate-${namespace}`;
    wrappedComponentCache.set(namespace, wrapped);
  }
  return wrapped;
}

export interface TranslateProps {
  namespace: string;
  translationKey: string;
  args?: { [key: string]: string };
  fallback?: ReactNode;
}

export default function Translate({ namespace, translationKey, args, fallback }: TranslateProps) {
  const { provider } = useContext(translationResourceProviderContext);

  // StaticTranslationProvider doesn't have a provider, render TranslateInner directly.
  if (provider === null) {
    return (
      <TranslateInner
        namespace={namespace}
        translationKey={translationKey}
        args={args}
        fallback={fallback}
      />
    );
  }

  const WrappedComponent = getOrCreateWrapped(namespace);
  return (
    <WrappedComponent
      namespace={namespace}
      translationKey={translationKey}
      args={args}
      fallback={fallback}
    />
  );
}
