// @rbx/not-approved-page-ui@2.0.0/dist/index.js
// src/providers/NotApprovedUIProvider.tsx
import { createContext, useContext } from "react";
import { jsx } from "react/jsx-runtime";
var NotApprovedUIContext = createContext(void 0);
var NotApprovedUIProvider = ({ config: config2, children }) => {
  return /* @__PURE__ */ jsx(NotApprovedUIContext.Provider, { value: config2, children });
};
var useNotApprovedUIConfig = () => {
  const config2 = useContext(NotApprovedUIContext);
  if (!config2) {
    throw new Error("useNotApprovedUIConfig must be used within a NotApprovedUIProvider");
  }
  return config2;
};
var useNotApprovedTranslate = () => {
  return useNotApprovedUIConfig().translate;
};

// src/containers/NotApprovedPageContainer.tsx
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";

// src/components/NotApprovedMainPage.tsx
import { Fragment as Fragment5, useEffect as useEffect5, useMemo as useMemo4, useState as useState12 } from "react";
import { Dialog as Dialog2, DialogTitle as DialogTitle4, DialogContent as DialogContent2, DialogBody as DialogBody4 } from "@rbx/foundation-ui";
import { EventTypes as EventTypes13 } from "@rbx/not-approved-page-events";

// src/services/useSendNotApprovedPageEvent.ts
import { useCallback } from "react";

// src/services/sendNotApprovedPageEvent.ts
import {
  EventContext,
  EVENT_NAME
} from "@rbx/not-approved-page-events";
var sendNotApprovedPageEvent = (sendAnalyticsEvent, platform, eventType, additionalProperties) => {
  const properties = {
    eventType,
    timestamp: Date.now(),
    platform,
    ...additionalProperties
  };
  sendAnalyticsEvent({
    eventName: EVENT_NAME,
    context: EventContext.NotApprovedPage,
    properties
  });
};
var sendNotApprovedPageEvent_default = sendNotApprovedPageEvent;

// src/services/useSendNotApprovedPageEvent.ts
function useSendNotApprovedPageEvent() {
  const { sendAnalyticsEvent, platform } = useNotApprovedUIConfig();
  return useCallback(
    (eventType, additionalProperties) => {
      sendNotApprovedPageEvent_default(sendAnalyticsEvent, platform, eventType, additionalProperties);
    },
    [sendAnalyticsEvent, platform]
  );
}

// src/services/useNotApprovedPageIxp.ts
import { useQuery } from "@tanstack/react-query";
var IXP_LAYER_NAME = "UserSafety.NotApprovedPage.UserID";
var useNotApprovedPageIxp = ({
  enabled
}) => {
  const { ixp } = useNotApprovedUIConfig();
  return useQuery({
    queryKey: [`ixp/${IXP_LAYER_NAME}`],
    queryFn: async () => {
      if (!ixp) return {};
      try {
        return await ixp.fetchLayer(IXP_LAYER_NAME);
      } catch (_e) {
        return {};
      }
    },
    // IXP data shouldn't change frequently so we can cache it indefinitely
    staleTime: Infinity,
    enabled: enabled && Boolean(ixp)
  });
};
var useNotApprovedPageIxp_default = useNotApprovedPageIxp;

// src/context/PageNavigationContext.tsx
import {
  createContext as createContext2,
  useContext as useContext2,
  useMemo,
  useState,
  useCallback as useCallback2
} from "react";
import { jsx as jsx2 } from "react/jsx-runtime";
var PageNavigationContext = createContext2(void 0);
var PageNavigationProvider = ({
  pages,
  unmappedViolationKeys,
  hasEducationalPages,
  commutationEligibility,
  children
}) => {
  const [page, setPage] = useState(0);
  const [hasNavigatedForward, setHasNavigatedForward] = useState(false);
  const currentPageName = pages[page]?.pageName ?? "unknown" /* Unknown */;
  const totalPages = pages.length;
  const isFirstPage = page === 0;
  const isLastPage = page === totalPages - 1;
  const isEducationalPassEligible = useMemo(
    () => commutationEligibility?.educational_pass_eligible ?? false,
    [commutationEligibility]
  );
  const goToNextPage = useCallback2(() => {
    if (page < totalPages - 1) {
      setHasNavigatedForward(true);
      setPage((prevPage) => prevPage + 1);
    }
  }, [page, totalPages]);
  const goToPreviousPage = useCallback2(() => {
    if (page > 0) {
      setPage((prevPage) => prevPage - 1);
    }
  }, [page]);
  const getProgress = useCallback2(() => {
    if (totalPages <= 1 || currentPageName === "second-chance-intro" /* SecondChanceIntro */ || currentPageName === "intro" /* Intro */) {
      return 0;
    }
    let totalPagesForProgress = totalPages - 1;
    let currentPageIndex = page;
    totalPagesForProgress -= isEducationalPassEligible ? 1 : 0;
    currentPageIndex -= isEducationalPassEligible ? 1 : 0;
    if (totalPagesForProgress <= 0) return 0;
    const progress = currentPageIndex / totalPagesForProgress * 100;
    return Math.round(progress);
  }, [page, totalPages, currentPageName, isEducationalPassEligible]);
  const contextValue = useMemo(
    () => ({
      currentPage: page,
      currentPageName,
      totalPages,
      goToNextPage,
      goToPreviousPage,
      isFirstPage,
      isLastPage,
      getProgress,
      hasEducationalPages,
      currentPageConfigs: pages[page]?.pageItems ?? [],
      CurrentCtaComponent: pages[page]?.CtaComponent,
      unmappedViolationKeys,
      hasNavigatedForward
    }),
    [
      page,
      currentPageName,
      totalPages,
      goToNextPage,
      goToPreviousPage,
      isFirstPage,
      isLastPage,
      getProgress,
      pages,
      hasEducationalPages,
      unmappedViolationKeys,
      hasNavigatedForward
    ]
  );
  return /* @__PURE__ */ jsx2(PageNavigationContext.Provider, { value: contextValue, children });
};
var usePageNavigation = () => {
  const context = useContext2(PageNavigationContext);
  if (!context) {
    throw new Error("usePageNavigation must be used within PageNavigationProvider");
  }
  return context;
};

// src/pageItemConfigs/PageItemsFromConfigs.tsx
import { jsx as jsx3 } from "react/jsx-runtime";
var PageItemsFromConfigs = ({ punishmentData, commutationEligibility }) => {
  const { isFirstPage, currentPageConfigs, currentPageName } = usePageNavigation();
  const pageItemsContent = currentPageConfigs.map((config2) => {
    if (!config2.getIsVisible(punishmentData, currentPageName, commutationEligibility)) {
      return null;
    }
    const PageItemComponent = config2.renderComponent;
    return /* @__PURE__ */ jsx3("div", { children: /* @__PURE__ */ jsx3(PageItemComponent, { punishmentData }) }, config2.configName);
  }).filter(Boolean);
  return /* @__PURE__ */ jsx3("div", { className: `flex flex-col ${!isFirstPage ? "gap-[36px]" : "gap-xlarge"}`, children: pageItemsContent });
};
var PageItemsFromConfigs_default = PageItemsFromConfigs;

// src/utils/constants.ts
var PUNISHMENT_TYPE = /* @__PURE__ */ ((PUNISHMENT_TYPE2) => {
  PUNISHMENT_TYPE2["Warn"] = "Warn";
  PUNISHMENT_TYPE2["Delete"] = "Delete";
  PUNISHMENT_TYPE2["Hour1"] = "Ban 1 Hour";
  PUNISHMENT_TYPE2["Hour6"] = "Ban 6 Hours";
  PUNISHMENT_TYPE2["Day1"] = "Ban 1 Day";
  PUNISHMENT_TYPE2["Day3"] = "Ban 3 Days";
  PUNISHMENT_TYPE2["Day7"] = "Ban 7 Days";
  PUNISHMENT_TYPE2["Day14"] = "Ban 14 Days";
  PUNISHMENT_TYPE2["Day60"] = "Ban 60 Days";
  PUNISHMENT_TYPE2["Month6"] = "Ban 6 Months";
  PUNISHMENT_TYPE2["Year1"] = "Ban 1 Year";
  return PUNISHMENT_TYPE2;
})(PUNISHMENT_TYPE || {});
var PUNISHMENT_TYPE_TO_STRING_KEY = {
  "Ban 1 Hour": "Heading.SuspendedOneHour",
  "Ban 6 Hours": "Heading.SuspendedSixHours",
  "Ban 1 Day": "Heading.SuspendedOneDay",
  "Ban 3 Days": "Heading.SuspendedThreeDays",
  "Ban 7 Days": "Heading.SuspendedSevenDays",
  "Ban 14 Days": "Heading.Suspended14Days",
  "Ban 60 Days": "Heading.SuspendedSixtyDays",
  "Ban 6 Months": "Heading.SuspendedSixMonths",
  "Ban 1 Year": "Heading.Suspended1Year",
  Warn: "Heading.Warning",
  Delete: "Heading.Banned"
};
var VIOLATION_TYPE_TO_PLURAL_MAP = {
  "Label.Type.Avatar": "Label.TypePlural.Avatar",
  "Label.Type.Audio": "Label.TypePlural.Audio",
  "Label.Type.AvatarAccessory": "Label.TypePlural.AvatarAccessory",
  "Label.Type.Bundle": "Label.TypePlural.Bundle",
  "Label.Type.Chat": "Label.TypePlural.Chat",
  "Label.Type.CommerceProduct": "Label.TypePlural.CommerceProduct",
  "Label.Type.Experience": "Label.TypePlural.Experience",
  "Label.Type.Game": "Label.TypePlural.Game",
  // same as experience violation, but updating external terminology.
  "Label.Type.Image": "Label.TypePlural.Image",
  "Label.Type.Look": "Label.TypePlural.Look",
  "Label.Type.Mesh": "Label.TypePlural.Mesh",
  "Label.Type.Model": "Label.TypePlural.Model",
  "Label.Type.Plugin": "Label.TypePlural.Plugin",
  "Label.Type.Video": "Label.TypePlural.Video",
  "Label.Type.Voice": "Label.TypePlural.Voice"
};
var REACTIVATION_CACHE_UPDATE_WAIT = 2e3;
var COMMUNITY_STANDARDS_URL = "https://en.help.roblox.com/hc/en-us/articles/203313410-Roblox-Community-Standards";
var UGC_GUIDELINES_URL = "https://create.roblox.com/docs/marketplace/marketplace-policy#general-creation-guidelines";
var ACCESS_MANAGEMENT_UPSELL_CAN_LIFT_PUNISHMENT_FEATURE_NAME = "CanRequestPunishmentLifting";
var VERIFICATION_CATEGORIES = {
  Email: "Email",
  VPC: "VPC",
  None: ""
};

// src/pageItemConfigs/educationalConfigs/policyEducationContentRegistry.ts
var SWEARING_POLICY_EDUCATION_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.Swearing",
  ruleSubtitle: "SubHeading.RuleExplanation.Swearing",
  ruleDescription: "Description.RuleExplanation.Swearing",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.Swearing",
  policyKey: "swearing"
};
var DATING_POLICY_EDUCATION_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.Dating",
  ruleSubtitle: "SubHeading.RuleExplanation.Dating",
  ruleDescription: "Description.RuleExplanation.Dating",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.Dating",
  policyKey: "dating"
};
var BULLYING_HARASSMENT_DISCRIMINATION_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.BullyingHarassmentDiscrimination",
  ruleSubtitle: "SubHeading.RuleExplanation.BullyingHarassmentDiscrimination",
  ruleDescription: "Description.RuleExplanation.BullyingHarassmentDiscrimination",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.BullyingHarassmentDiscrimination",
  policyKey: "bullying-harassment-discrimination"
};
var DIRECTING_USERS_OFF_PLATFORM_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.DirectingUsersOffPlatform",
  ruleSubtitle: "SubHeading.RuleExplanation.DirectingUsersOffPlatform",
  ruleDescription: "Description.RuleExplanation.DirectingUsersOffPlatform",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.DirectingUsersOffPlatform",
  policyKey: "directing-users-off-platform"
};
var MISUSING_ROBLOX_SYSTEMS_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.MisusingRobloxSystems",
  ruleSubtitle: "SubHeading.RuleExplanation.MisusingRobloxSystems",
  ruleDescription: "Description.RuleExplanation.MisusingRobloxSystems",
  ruleDescriptionBullets: "Description.RuleExplanation.Bullets.MisusingRobloxSystems",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.MisusingRobloxSystems",
  policyKey: "misusing-roblox-systems"
};
var PII_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.PII",
  ruleSubtitle: "SubHeading.RuleExplanation.PII",
  ruleDescription: "Description.RuleExplanation.PII",
  ruleDescriptionBullets: "Description.RuleExplanation.Bullets.PII",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.PII",
  policyKey: "pii"
};
var REAL_LIFE_EVENTS_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.RealLifeEvents",
  ruleSubtitle: "SubHeading.RuleExplanation.RealLifeEvents",
  ruleDescription: "Description.RuleExplanation.RealLifeEvents",
  ruleDescriptionBullets: "Description.RuleExplanation.Bullets.RealLifeEvents",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.RealLifeEvents",
  policyKey: "real-life-events"
};
var SPAM_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.Spam",
  ruleSubtitle: "SubHeading.RuleExplanation.Spam",
  ruleDescription: "Description.RuleExplanation.Spam",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.Spam",
  policyKey: "spam"
};
var SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.SSHAndIllegalRegulatedActivities",
  ruleSubtitle: "SubHeading.RuleExplanation.SSHAndIllegalRegulatedActivities",
  ruleDescription: "Description.RuleExplanation.SSHAndIllegalRegulatedActivities",
  // The bullets key is different because it's too long for Translations Hub.
  ruleDescriptionBullets: "Description.RuleExplanation.Bullets.SSHAndIllegalRegAct",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.SSHAndIllegalRegulatedActivities",
  policyKey: "ssh-and-illegal-regulated-activities"
};
var VIOLENT_CONTENT_AND_GORE_CONTENT = {
  ruleTitle: "Heading.RuleExplanation.ViolentContentAndGore",
  ruleSubtitle: "SubHeading.RuleExplanation.ViolentContentAndGore",
  ruleDescription: "Description.RuleExplanation.ViolentContentAndGore",
  importanceTitle: "Heading.RuleImportance",
  importanceDescription: "Description.RuleImportance.ViolentContentAndGore",
  policyKey: "violent-content-and-gore"
};
var POLICY_EDUCATION_CONTENT_REGISTRY = {
  "Label.AbuseType.CheatandExploits": MISUSING_ROBLOX_SYSTEMS_CONTENT,
  "Label.AbuseType.ContestsandSweepstakes": SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT,
  "Label.AbuseType.Dating": DATING_POLICY_EDUCATION_CONTENT,
  "Label.AbuseType.DirectingUsersOffPlatform": DIRECTING_USERS_OFF_PLATFORM_CONTENT,
  "Label.AbuseType.DiscriminatoryContent": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.DisruptiveAudio": MISUSING_ROBLOX_SYSTEMS_CONTENT,
  "Label.AbuseType.EncouragingDangerousBehavior": SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT,
  "Label.AbuseType.ExtortionandBlackmail": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.Harassment": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.IllegalandRegulatedContent": SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT,
  "Label.AbuseType.Impersonation": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.IrlDangerousActivities": SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT,
  "Label.AbuseType.MisusingRobloxSystems": MISUSING_ROBLOX_SYSTEMS_CONTENT,
  "Label.AbuseType.OffPlatformSpeechandBehavior": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.PrivacyAskingforPII": PII_CONTENT,
  "Label.AbuseType.PrivacyGivingPII": PII_CONTENT,
  "Label.AbuseType.RealLifeThreats": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.Scamming": MISUSING_ROBLOX_SYSTEMS_CONTENT,
  "Label.AbuseType.SexualContent": DATING_POLICY_EDUCATION_CONTENT,
  "Label.AbuseType.Spam": SPAM_CONTENT,
  "Label.AbuseType.SuicideSelfHarm": SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT,
  "Label.AbuseType.Swearing": SWEARING_POLICY_EDUCATION_CONTENT,
  "Label.AbuseType.ThreatsOrAbuseOfRobloxEmployeesOrAffiliates": BULLYING_HARASSMENT_DISCRIMINATION_CONTENT,
  "Label.AbuseType.ViolentContentAndGore": VIOLENT_CONTENT_AND_GORE_CONTENT,
  "Label.AbuseType.VirtualCasino": SSH_AND_ILLEGAL_REGULATED_ACTIVITIES_CONTENT,
  "Label.Sublabel.RealLifeEvents": REAL_LIFE_EVENTS_CONTENT,
  "Label.Sublabel.RomanceOrSex": DATING_POLICY_EDUCATION_CONTENT
};
var policyEducationContentRegistry_default = POLICY_EDUCATION_CONTENT_REGISTRY;

// src/utils/makeTranslateWithLink.tsx
import { jsx as jsx4, jsxs } from "react/jsx-runtime";
var START_LINK_PLACEHOLDER = "{startLink}";
var END_LINK_PLACEHOLDER = "{endLink}";
var makeTranslateWithLink = (translate, sendEvent) => {
  const translateWithLink = (key, href, params, eventStreamEventName) => {
    const translated = translate(key, {
      startLink: START_LINK_PLACEHOLDER,
      endLink: END_LINK_PLACEHOLDER,
      ...params
    });
    const [beforeLink, withinAndAfterLink] = translated.split(START_LINK_PLACEHOLDER, 2);
    const [withinLink, afterLink] = withinAndAfterLink?.split(END_LINK_PLACEHOLDER, 2) ?? [];
    return /* @__PURE__ */ jsxs("p", { className: "text-body-large", children: [
      beforeLink,
      /* @__PURE__ */ jsx4(
        "a",
        {
          href,
          className: "content-link",
          rel: "noreferrer",
          onClick: () => {
            if (eventStreamEventName != null && sendEvent) {
              sendEvent(eventStreamEventName);
            }
          },
          children: withinLink
        }
      ),
      afterLink
    ] });
  };
  return translateWithLink;
};
var makeTranslateWithLink_default = makeTranslateWithLink;

// src/pageItemConfigs/configs/ChargebackStepsPageItemConfig.tsx
import { jsx as jsx5, jsxs as jsxs2 } from "react/jsx-runtime";
var ChargebackStepsPageItem = ({ punishmentData }) => {
  const translate = useNotApprovedTranslate();
  const { websiteUrl } = useNotApprovedUIConfig();
  const translateWithLink = makeTranslateWithLink_default(translate);
  const { verificationCategory } = punishmentData;
  const homePageUrl = `${websiteUrl}/`;
  const reactivationContent = verificationCategory === VERIFICATION_CATEGORIES.VPC ? translateWithLink("Label.ParentReactivationNotice", homePageUrl) : translateWithLink("Label.EmailReactivationNotice", homePageUrl);
  return /* @__PURE__ */ jsxs2("div", { className: "flex flex-col gap-medium", children: [
    /* @__PURE__ */ jsx5("span", { className: "text-heading-medium", children: translate("Label.ChargebackNextSteps") }),
    reactivationContent
  ] });
};
var ChargebackStepsPageItemConfig = {
  getIsVisible: (punishmentData) => [VERIFICATION_CATEGORIES.Email, VERIFICATION_CATEGORIES.VPC].includes(
    punishmentData.verificationCategory
  ),
  renderComponent: ChargebackStepsPageItem,
  configName: "chargeback-steps"
};
var ChargebackStepsPageItemConfig_default = ChargebackStepsPageItemConfig;

// src/pageItemConfigs/configs/PreventionStepsPageItemConfig.tsx
import { Button } from "@rbx/foundation-ui";
import { EventTypes as EventTypes2 } from "@rbx/not-approved-page-events";
import { jsx as jsx6, jsxs as jsxs3 } from "react/jsx-runtime";
var PreventionStepsPageItem = ({ punishmentData }) => {
  const translate = useNotApprovedTranslate();
  const sendEvent = useSendNotApprovedPageEvent();
  const { hasEducationalPages } = usePageNavigation();
  const { showUGCAvatarGuidelinesLink, context } = punishmentData;
  const handleViewRules = (showUGC) => {
    sendEvent(showUGC ? EventTypes2.UGCGuidelinesClicked : EventTypes2.CommunityGuidelineClicked);
    window.open(showUGC ? UGC_GUIDELINES_URL : COMMUNITY_STANDARDS_URL, "_blank");
  };
  const showRulesButton = Boolean(showUGCAvatarGuidelinesLink) || !hasEducationalPages;
  const isAltInformed = context?.IS_ALT_INFORMED;
  return /* @__PURE__ */ jsxs3("div", { className: "flex flex-col gap-medium", children: [
    /* @__PURE__ */ jsx6("span", { className: "text-heading-medium", children: translate("Label.RuleBreakingAddsUp") }),
    /* @__PURE__ */ jsxs3("div", { className: "flex flex-col gap-medium", children: [
      /* @__PURE__ */ jsx6("p", { className: "text-body-large", children: translate("Description.Foreshadow") }),
      isAltInformed && /* @__PURE__ */ jsx6("p", { className: "text-body-large", children: translate("Description.LinkedAccounts") }),
      showRulesButton && /* @__PURE__ */ jsx6(
        Button,
        {
          variant: "Standard",
          size: "Small",
          onClick: () => {
            handleViewRules(showUGCAvatarGuidelinesLink ?? false);
          },
          children: showUGCAvatarGuidelinesLink ? translate("Action.ViewRulesUGCAvatarGuidelines") : translate("Action.ViewRules")
        }
      )
    ] })
  ] });
};
var PreventionStepsPageItemConfig = {
  getIsVisible: (punishmentData) => {
    return punishmentData.punishmentTypeDescription !== "Delete" /* Delete */;
  },
  renderComponent: PreventionStepsPageItem,
  configName: "prevention-steps"
};
var PreventionStepsPageItemConfig_default = PreventionStepsPageItemConfig;

// src/utils/determinePunishmentDescription.ts
function determinePunishmentDescription(violationReasons, violationType, interventionType, translate) {
  if (interventionType === "Delete" /* Delete */) {
    return translate("Description.BrokeRulesBanned");
  }
  const pluralViolationType = VIOLATION_TYPE_TO_PLURAL_MAP[violationType ?? ""];
  const translatedViolationType = pluralViolationType ? translate(pluralViolationType) : "";
  if (violationReasons.length > 0) {
    return translatedViolationType ? translate("Description.BrokeRulesTypePolicy.V2", {
      type: translatedViolationType,
      policy: violationReasons.join(", ").toLowerCase()
    }) : translate("Description.BrokeRulesPolicy.V2", {
      policy: violationReasons.join(", ").toLowerCase()
    });
  }
  return translatedViolationType ? translate("Description.BrokeRulesType", {
    type: translatedViolationType
  }) : translate("Description.BrokeRulesGeneric");
}
var determinePunishmentDescription_default = determinePunishmentDescription;

// src/context/NotApprovedPagePunishmentProvider.tsx
import { createContext as createContext3, useContext as useContext3, useMemo as useMemo2 } from "react";

// src/services/usePunishmentData.ts
import { useQuery as useQuery2 } from "@tanstack/react-query";
var usePunishmentData = () => {
  const { httpGet, userModerationApiUrl } = useNotApprovedUIConfig();
  return useQuery2({
    queryKey: ["not-approved-data"],
    queryFn: () => httpGet(`${userModerationApiUrl}/v1/not-approved`),
    /**
     * We don't anticipate the punishment data changing for a user dynamically at all so we can set the stale
     * time to Infinity to never refetch the data. The only time the data can change is when the user refreshes
     * the page.
     */
    staleTime: Infinity
  });
};
var usePunishmentData_default = usePunishmentData;

// src/utils/platformEvidenceTypes.ts
import * as z from "zod/mini";
z.config(z.locales.en());
var zPlatformEvidence = z.object({
  type: z.literal("platform" /* PLATFORM */),
  displayMeta: z.optional(
    z.object({
      lowercaseKey: z.string().check(z.minLength(1)),
      capitalizedKey: z.string().check(z.minLength(1)),
      icon: z.string()
    })
  ),
  /** We handle this verification separately in `zPlatformElement` */
  elements: z.optional(z.array(z.unknown()))
});
var isPlatformEvidence = (evidence) => {
  if (evidence.type !== "platform" /* PLATFORM */) {
    return false;
  }
  const result = zPlatformEvidence.safeParse(evidence);
  if (!result.success) {
    console.warn("Failed to parse platform evidence: ", result.error);
    return false;
  }
  return true;
};
var zPlatformElement = z.union([
  z.object({
    type: z.literal("text"),
    labelKey: z.string(),
    text: z.optional(z.string()),
    textKey: z.optional(z.string()),
    textKeyParameters: z.optional(z.record(z.string(), z.string()))
  }).check(
    z.refine((data) => data.text !== void 0 || data.textKey !== void 0, {
      message: "Either 'text' or 'textKey' must be provided"
    }),
    z.refine((data) => !data.textKeyParameters || data.textKey !== void 0, {
      message: "'textKeyParameters' requires 'textKey' to be provided"
    })
  ),
  z.object({
    type: z.literal("image"),
    labelKey: z.string(),
    url: z.string(),
    altLabelKey: z.optional(z.string())
  })
  // TODO(agrossman): allow rendering of all supported elements
  // z.object({
  //   type: z.literal('timestamp'),
  //   labelKey: z.string(),
  //   unix: z.number()
  // })
]);
var isPlatformElementValid = (element) => {
  const result = zPlatformElement.safeParse(element);
  if (!result.success) {
    console.warn("Failed to parse platform element: ", result.error);
    return false;
  }
  return true;
};
var zPlatformEvidenceFullyTyped = z.object({
  ...zPlatformEvidence.shape,
  elements: z.optional(z.array(zPlatformElement))
});
var isValidatedPlatformEvidence = (evidence) => {
  if (evidence.type !== "platform" /* PLATFORM */) {
    return false;
  }
  const result = zPlatformEvidenceFullyTyped.safeParse(evidence);
  if (!result.success) {
    console.warn("Failed to parse platform evidence fully typed: ", result.error);
    return false;
  }
  return true;
};

// src/utils/isPlatformEvidenceVisibleInView.ts
var isPlatformEvidenceVisibleInView = (punishmentData) => {
  const violationData = punishmentData.violation;
  if (!violationData) {
    return false;
  }
  const { evidence } = violationData;
  if (!evidence || !isPlatformEvidence(evidence)) {
    return false;
  }
  const visibleElements = evidence.elements?.filter((element) => {
    return isPlatformElementValid(element);
  });
  if (!visibleElements || visibleElements.length === 0) {
    return false;
  }
  return isValidatedPlatformEvidence({
    ...evidence,
    elements: visibleElements
  });
};
var isPlatformEvidenceVisibleInView_default = isPlatformEvidenceVisibleInView;

// src/services/useCommutationEligibility.ts
import { useQuery as useQuery3 } from "@tanstack/react-query";
var useCommutationEligibility = () => {
  const { httpGet, apiGatewayUrl } = useNotApprovedUIConfig();
  return useQuery3({
    queryKey: ["commutation-eligibility"],
    queryFn: async () => {
      try {
        return await httpGet(
          `${apiGatewayUrl}/moderation-appeal-service/v2/consequence-commutation-eligibility`
        );
      } catch (_error) {
        return {
          educational_pass_eligible: false
        };
      }
    },
    staleTime: 5 * 60 * 1e3
    // 5 minutes
  });
};
var useCommutationEligibility_default = useCommutationEligibility;

// src/context/NotApprovedPagePunishmentProvider.tsx
import { jsx as jsx7 } from "react/jsx-runtime";
var NotApprovedPagePunishmentContext = createContext3(void 0);
var NotApprovedPagePunishmentProvider = ({
  enableIxp = false,
  children
}) => {
  const translate = useNotApprovedTranslate();
  const { data: rawPunishmentData, isLoading, error } = usePunishmentData_default();
  const { data: ixpData, isLoading: isLoadingIxp } = useNotApprovedPageIxp_default({ enabled: enableIxp });
  const { data: commutationEligibility, isLoading: isLoadingCommutation } = useCommutationEligibility_default();
  const punishmentData = rawPunishmentData?.punishedUserId ? rawPunishmentData : void 0;
  const violationReasons = useMemo2(() => {
    const translatedReasons = /* @__PURE__ */ new Set();
    const untranslatedReasons = /* @__PURE__ */ new Set();
    if (punishmentData?.violation && isPlatformEvidenceVisibleInView_default(punishmentData)) {
      punishmentData.violation.abuseTypeTranslationKeys.forEach((key) => {
        translatedReasons.add(translate(key));
        untranslatedReasons.add(key);
      });
    } else {
      punishmentData?.badUtterances?.forEach((utterance) => {
        translatedReasons.add(translate(utterance.labelTranslationKey));
        untranslatedReasons.add(utterance.labelTranslationKey);
      });
    }
    return {
      translatedReasons: [...translatedReasons].filter(Boolean),
      untranslatedReasons: [...untranslatedReasons].filter(Boolean)
    };
  }, [punishmentData, translate]);
  const aggregatedIsLoading = isLoading || enableIxp && isLoadingIxp || isLoadingCommutation;
  const contextValue = useMemo2(
    () => ({
      isLoading: aggregatedIsLoading,
      error,
      isReady: !aggregatedIsLoading && !error && punishmentData !== void 0,
      punishmentData,
      violationReasons,
      ixpData,
      commutationEligibility
    }),
    [aggregatedIsLoading, error, punishmentData, violationReasons, ixpData, commutationEligibility]
  );
  return /* @__PURE__ */ jsx7(NotApprovedPagePunishmentContext.Provider, { value: contextValue, children });
};
var useNotApprovedPagePunishment = () => {
  const context = useContext3(NotApprovedPagePunishmentContext);
  if (!context) {
    throw new Error(
      "useNotApprovedPagePunishment must be used within a NotApprovedPagePunishmentProvider"
    );
  }
  return context;
};

// src/pageItemConfigs/configs/PunishmentDescriptionPageItemConfig.tsx
import { jsx as jsx8 } from "react/jsx-runtime";
var PunishmentDescription = ({ punishmentData }) => {
  const translate = useNotApprovedTranslate();
  const { violationReasons } = useNotApprovedPagePunishment();
  const { violation, punishmentTypeDescription } = punishmentData;
  const violationType = violation?.evidence?.displayMeta?.capitalizedKey;
  const punishmentDescription = determinePunishmentDescription_default(
    violationReasons?.translatedReasons ?? [],
    violationType,
    punishmentTypeDescription,
    translate
  );
  return /* @__PURE__ */ jsx8("span", { className: "text-body-medium", children: punishmentDescription });
};
var PunishmentDescriptionPageItemConfig = {
  getIsVisible: () => true,
  renderComponent: PunishmentDescription,
  configName: "punishment-description"
};
var PunishmentDescriptionPageItemConfig_default = PunishmentDescriptionPageItemConfig;

// src/pageItemConfigs/configs/ReportMistakePageItemConfig.tsx
import { Button as Button2 } from "@rbx/foundation-ui";
import { EventTypes as EventTypes3 } from "@rbx/not-approved-page-events";
import { jsx as jsx9, jsxs as jsxs4 } from "react/jsx-runtime";
var ReportMistakePageItem = () => {
  const translate = useNotApprovedTranslate();
  const sendEvent = useSendNotApprovedPageEvent();
  const { websiteUrl } = useNotApprovedUIConfig();
  const handleReportMistakeButtonClick = () => {
    sendEvent(EventTypes3.AppealsPortalClicked);
    window.open(`${websiteUrl}/report-appeals?t_source=nap-web`, "_blank");
  };
  return /* @__PURE__ */ jsxs4("div", { className: "flex flex-col gap-medium", children: [
    /* @__PURE__ */ jsx9("p", { className: "text-body-large", children: translate("Description.Mistake.V3") }),
    /* @__PURE__ */ jsx9(Button2, { variant: "Standard", size: "Small", onClick: handleReportMistakeButtonClick, children: translate("Action.SendAppeal") })
  ] });
};
var ReportMistakePageItemConfig = {
  getIsVisible: (_, pageName, commutationEligibility) => {
    const isEducationalPassEligible = commutationEligibility?.educational_pass_eligible ?? false;
    return !isEducationalPassEligible || pageName === "second-chance-intro" /* SecondChanceIntro */;
  },
  renderComponent: ReportMistakePageItem,
  configName: "report-mistake"
};
var ReportMistakePageItemConfig_default = ReportMistakePageItemConfig;

// src/utils/safeTranslateWithParams.ts
import { EventTypes as EventTypes4 } from "@rbx/not-approved-page-events";
var safeTranslateWithParams = (translate, key, params, sendEvent) => {
  if (!params) return translate(key);
  try {
    return translate(key, params);
  } catch {
    sendEvent?.(EventTypes4.MissingTranslation, {
      additionalInfo: JSON.stringify({ key, params })
    });
    return translate(key);
  }
};
var safeTranslateWithParams_default = safeTranslateWithParams;

// src/components/ImageWithSpinner.tsx
import { useState as useState2 } from "react";
import { ProgressCircle } from "@rbx/foundation-ui";
import { jsx as jsx10, jsxs as jsxs5 } from "react/jsx-runtime";
var ImageWithSpinner = ({ url, altLabelKey }) => {
  const translate = useNotApprovedTranslate();
  const [loading, setLoading] = useState2(true);
  return /* @__PURE__ */ jsxs5("div", { className: "flex aspect-1-1 justify-center items-center bg-over-media-100 radius-medium", children: [
    /* @__PURE__ */ jsx10(
      "img",
      {
        style: { objectFit: "contain" },
        className: `size-full ${loading ? "hidden" : ""}`,
        src: url,
        alt: altLabelKey && translate(altLabelKey),
        onLoad: () => {
          setLoading(false);
        },
        onError: () => {
          setLoading(false);
        }
      }
    ),
    /* @__PURE__ */ jsx10("div", { "data-testid": "spinner-container", className: loading ? void 0 : "hidden", children: /* @__PURE__ */ jsx10(
      ProgressCircle,
      {
        ariaLabel: translate("Label.LoadingImage"),
        size: "Medium",
        variant: "Indeterminate"
      }
    ) })
  ] });
};
var ImageWithSpinner_default = ImageWithSpinner;

// src/utils/getFormattedFullDate.ts
var defaultFormatFullDate = (date) => {
  const d = new Date(date);
  return d.toLocaleString(void 0, {
    month: "long",
    day: "numeric",
    year: "numeric",
    hour: "numeric",
    minute: "numeric"
  });
};
function useFormatFullDate() {
  const { formatFullDate } = useNotApprovedUIConfig();
  return formatFullDate ?? defaultFormatFullDate;
}

// src/components/EvidenceField.tsx
import { jsx as jsx11, jsxs as jsxs6 } from "react/jsx-runtime";
var EvidenceField = ({ fieldLabel, fieldValue, preline = false }) => {
  const labelWithColon = `${fieldLabel}:`;
  return /* @__PURE__ */ jsxs6("div", { className: "flex flex-col", children: [
    /* @__PURE__ */ jsx11("span", { className: "text-title-medium", children: labelWithColon }),
    /* @__PURE__ */ jsx11(
      "p",
      {
        className: "text-body-medium",
        style: {
          wordBreak: "break-word",
          whiteSpace: preline ? "pre-line" : "normal"
        },
        children: fieldValue
      }
    )
  ] });
};
var EvidenceField_default = EvidenceField;

// src/components/ViewActivityButton.tsx
import { Button as Button3 } from "@rbx/foundation-ui";
import { EventTypes as EventTypes5 } from "@rbx/not-approved-page-events";
import { jsx as jsx12 } from "react/jsx-runtime";
var ViewActivityButton = () => {
  const translate = useNotApprovedTranslate();
  const sendEvent = useSendNotApprovedPageEvent();
  const { websiteUrl } = useNotApprovedUIConfig();
  const handleViewAllActivity = () => {
    sendEvent(EventTypes5.AppealsPortalClicked);
    window.open(`${websiteUrl}/report-appeals?t_source=nap-web`, "_blank");
  };
  return /* @__PURE__ */ jsx12(Button3, { onClick: handleViewAllActivity, variant: "Standard", size: "Small", children: translate("Action.ViewMore") });
};
var ViewActivityButton_default = ViewActivityButton;

// src/pageItemConfigs/configs/ViolationPageItemConfig.tsx
import { jsx as jsx13, jsxs as jsxs7 } from "react/jsx-runtime";
var ViolationPageItem = ({ punishmentData }) => {
  const { violation, consequenceTransparencyMessage } = punishmentData;
  const translate = useNotApprovedTranslate();
  const formatFullDate = useFormatFullDate();
  const sendEvent = useSendNotApprovedPageEvent();
  const elements = violation?.evidence?.elements;
  const validElements = elements?.filter((element) => isPlatformElementValid(element)) ?? [];
  const imageElements = [];
  const textElements = [];
  validElements.forEach((elem) => {
    if (elem.type === "image") {
      imageElements.push(
        /* @__PURE__ */ jsx13(ImageWithSpinner_default, { url: elem.url, altLabelKey: elem.labelKey }, elem.url)
      );
      return;
    }
    textElements.push(
      /* @__PURE__ */ jsx13(
        EvidenceField_default,
        {
          fieldLabel: translate(elem.labelKey),
          fieldValue: elem.textKey ? safeTranslateWithParams_default(translate, elem.textKey, elem.textKeyParameters, sendEvent) : elem.text ?? ""
        },
        elem.labelKey
      )
    );
  });
  const isImageShown = imageElements.length > 0;
  return /* @__PURE__ */ jsxs7(
    "div",
    {
      "data-testid": "violation-grid",
      className: "flex flex-col gap-medium medium:flex-row medium:gap-large",
      children: [
        isImageShown && /* @__PURE__ */ jsx13("div", { className: "shrink-0 width-[50%] medium:width-full medium:basis-[40%]", children: imageElements[0] }),
        /* @__PURE__ */ jsxs7("div", { className: "flex flex-col gap-small width-full", children: [
          textElements,
          /* @__PURE__ */ jsx13(
            EvidenceField_default,
            {
              fieldLabel: translate("Label.ReviewDate"),
              fieldValue: formatFullDate(punishmentData.beginDate)
            }
          ),
          consequenceTransparencyMessage && /* @__PURE__ */ jsx13(
            EvidenceField_default,
            {
              fieldLabel: translate("Label.DecisionMethod"),
              fieldValue: consequenceTransparencyMessage
            }
          ),
          /* @__PURE__ */ jsx13(ViewActivityButton_default, {})
        ] })
      ]
    }
  );
};
var ViolationPageItemConfig = {
  getIsVisible: isPlatformEvidenceVisibleInView_default,
  renderComponent: ViolationPageItem,
  configName: "violation-evidence"
};
var ViolationPageItemConfig_default = ViolationPageItemConfig;

// src/pageItemConfigs/configs/BadUtterancesPageItemConfig.tsx
import { useState as useState3 } from "react";
import { Button as Button4, Divider } from "@rbx/foundation-ui";
import { jsx as jsx14, jsxs as jsxs8 } from "react/jsx-runtime";
var DEFAULT_NUM_UTTERANCES = 4;
var COLON = ":";
var BadUtterancesContent = ({ badUtterances }) => {
  const [showAll, setShowAll] = useState3(false);
  const translate = useNotApprovedTranslate();
  return /* @__PURE__ */ jsxs8("div", { className: "flex flex-col gap-small", children: [
    /* @__PURE__ */ jsxs8("div", { className: "flex flex-col", children: [
      /* @__PURE__ */ jsxs8("span", { className: "text-title-medium", children: [
        translate("Label.OffensiveItem.V2"),
        COLON
      ] }),
      /* @__PURE__ */ jsxs8("div", { className: "flex flex-col items-start", children: [
        badUtterances.slice(0, DEFAULT_NUM_UTTERANCES).map((element) => /* @__PURE__ */ jsx14(
          "p",
          {
            className: "text-body-medium",
            style: { wordBreak: "break-word" },
            children: element.utteranceText
          },
          element.utteranceText
        )),
        badUtterances.slice(DEFAULT_NUM_UTTERANCES).map((element) => /* @__PURE__ */ jsx14(
          "p",
          {
            className: `text-body-medium ${showAll ? "" : "hidden"}`,
            style: { wordBreak: "break-word" },
            children: element.utteranceText
          },
          element.utteranceText
        ))
      ] })
    ] }),
    badUtterances.length > DEFAULT_NUM_UTTERANCES && /* @__PURE__ */ jsx14(
      Button4,
      {
        className: "self-start margin-left-[-7px]",
        variant: "Link",
        size: "XSmall",
        onClick: () => {
          setShowAll((prev) => !prev);
        },
        "data-testid": "view-toggle-button",
        children: showAll ? translate("Action.ViewLess") : translate("Action.ViewMore")
      }
    )
  ] });
};
var BadUtterancesPageItem = ({ punishmentData }) => {
  const translate = useNotApprovedTranslate();
  const formatFullDate = useFormatFullDate();
  const { badUtterances, consequenceTransparencyMessage } = punishmentData;
  return /* @__PURE__ */ jsxs8("div", { className: "flex flex-col gap-small", "data-testid": "bad-utterances", children: [
    /* @__PURE__ */ jsx14(BadUtterancesContent, { badUtterances: badUtterances ?? [] }),
    /* @__PURE__ */ jsx14(Divider, {}),
    /* @__PURE__ */ jsx14(
      EvidenceField_default,
      {
        fieldLabel: translate("Label.ReviewDate"),
        fieldValue: formatFullDate(punishmentData.beginDate)
      }
    ),
    consequenceTransparencyMessage && /* @__PURE__ */ jsx14(
      EvidenceField_default,
      {
        fieldLabel: translate("Label.DecisionMethod"),
        fieldValue: consequenceTransparencyMessage
      }
    ),
    /* @__PURE__ */ jsx14(ViewActivityButton_default, {})
  ] });
};
var BadUtterancesPageItemConfig = {
  getIsVisible: (punishmentData) => {
    return !isPlatformEvidenceVisibleInView_default(punishmentData) && (punishmentData.badUtterances?.length ?? 0) > 0;
  },
  renderComponent: BadUtterancesPageItem,
  configName: "bad-utterances"
};
var BadUtterancesPageItemConfig_default = BadUtterancesPageItemConfig;

// src/pageItemConfigs/configs/SimpleEvidencePageItemConfig.tsx
import { jsx as jsx15, jsxs as jsxs9 } from "react/jsx-runtime";
var SimpleEvidencePageItem = (props) => {
  const translate = useNotApprovedTranslate();
  const formatFullDate = useFormatFullDate();
  const { punishmentData } = props;
  const { beginDate, consequenceTransparencyMessage } = punishmentData;
  return /* @__PURE__ */ jsxs9("div", { className: "flex flex-col gap-small", "data-testid": "simple-evidence", children: [
    /* @__PURE__ */ jsx15(
      EvidenceField_default,
      {
        fieldLabel: translate("Label.ReviewDate"),
        fieldValue: formatFullDate(beginDate)
      }
    ),
    consequenceTransparencyMessage && /* @__PURE__ */ jsx15(
      EvidenceField_default,
      {
        fieldLabel: translate("Label.DecisionMethod"),
        fieldValue: consequenceTransparencyMessage
      }
    ),
    /* @__PURE__ */ jsx15(ViewActivityButton_default, {})
  ] });
};
var SimpleEvidencePageItemConfig = {
  getIsVisible: (punishmentData) => {
    return !isPlatformEvidenceVisibleInView_default(punishmentData) && (punishmentData.badUtterances?.length ?? 0) === 0;
  },
  renderComponent: SimpleEvidencePageItem,
  configName: "simple-evidence"
};
var SimpleEvidencePageItemConfig_default = SimpleEvidencePageItemConfig;

// src/pageItemConfigs/reviewedEvidenceConfigs.ts
var getReviewedEvidencePageItemConfigs = () => {
  return [ViolationPageItemConfig_default, BadUtterancesPageItemConfig_default, SimpleEvidencePageItemConfig_default];
};

// src/pageItemConfigs/configs/ReviewedEvidencePageItemConfig.tsx
import { jsx as jsx16, jsxs as jsxs10 } from "react/jsx-runtime";
var ReviewedEvidencePageItem = ({ punishmentData }) => {
  const translate = useNotApprovedTranslate();
  const renderedPageItems = getReviewedEvidencePageItemConfigs().map((config2) => {
    return config2.getIsVisible(punishmentData) ? /* @__PURE__ */ jsx16("div", { children: config2.renderComponent({ punishmentData }) }, config2.configName) : null;
  }).filter(Boolean);
  return /* @__PURE__ */ jsxs10("div", { className: "flex flex-col gap-medium", "data-testid": "reviewed-evidence-container", children: [
    /* @__PURE__ */ jsx16("span", { className: "text-title-large", children: translate("Label.LatestActivity") }),
    /* @__PURE__ */ jsxs10("div", { className: "flex flex-col gap-xsmall", children: [
      /* @__PURE__ */ jsx16("div", { className: "padding-large bg-shift-200 radius-medium", children: renderedPageItems }),
      /* @__PURE__ */ jsx16("span", { className: "text-caption-medium content-muted", children: punishmentData.interventionId })
    ] })
  ] });
};
var ReviewedEvidencePageItemConfig = {
  getIsVisible: () => true,
  renderComponent: ReviewedEvidencePageItem,
  configName: "rewiewed-evidence"
};
var ReviewedEvidencePageItemConfig_default = ReviewedEvidencePageItemConfig;

// src/pageItemConfigs/secondChanceConfigs/SecondChanceConclusionPageItemConfig.tsx
import { jsx as jsx17, jsxs as jsxs11 } from "react/jsx-runtime";
var SecondChanceConclusionPageItem = () => {
  const translate = useNotApprovedTranslate();
  return /* @__PURE__ */ jsxs11("div", { className: "flex flex-col gap-medium", children: [
    /* @__PURE__ */ jsx17("span", { className: "text-heading-medium", children: translate("Heading.SecondChance.Details") }),
    /* @__PURE__ */ jsx17("p", { className: "text-body-large", children: translate("Description.SecondChance.Details") })
  ] });
};
var SecondChanceConclusionPageItemConfig = {
  getIsVisible: () => true,
  renderComponent: SecondChanceConclusionPageItem,
  configName: "second-chance-conclusion"
};
var SecondChanceConclusionPageItemConfig_default = SecondChanceConclusionPageItemConfig;

// src/pageItemConfigs/secondChanceConfigs/SecondChanceIntroPageItemConfig.tsx
import { jsx as jsx18, jsxs as jsxs12 } from "react/jsx-runtime";
var SecondChanceIntroPageItem = () => {
  const translate = useNotApprovedTranslate();
  return /* @__PURE__ */ jsxs12("div", { className: "flex flex-col gap-large", children: [
    /* @__PURE__ */ jsx18("span", { className: "text-heading-medium", children: translate("Heading.SecondChance") }),
    /* @__PURE__ */ jsx18("p", { className: "text-body-large", children: translate("Description.SecondChance.Second") })
  ] });
};
var SecondChanceIntroPageItemConfig = {
  getIsVisible: () => true,
  renderComponent: SecondChanceIntroPageItem,
  configName: "second-chance-intro"
};
var SecondChanceIntroPageItemConfig_default = SecondChanceIntroPageItemConfig;

// src/pageItemConfigs/configs/WhatHappenedPageItemConfig.tsx
import { jsx as jsx19, jsxs as jsxs13 } from "react/jsx-runtime";
var WhatHappenedPageItem = (props) => {
  const translate = useNotApprovedTranslate();
  const { violationReasons } = useNotApprovedPagePunishment();
  const { punishmentData } = props;
  const { messageToUser } = punishmentData;
  return /* @__PURE__ */ jsxs13("div", { className: "flex flex-col gap-medium", "data-testid": "what-happened", children: [
    /* @__PURE__ */ jsx19("span", { className: "text-title-large", children: translate("Label.WhatHappened") }),
    /* @__PURE__ */ jsxs13("div", { className: "padding-large bg-shift-200 radius-medium flex flex-col gap-small", children: [
      /* @__PURE__ */ jsx19(
        EvidenceField_default,
        {
          fieldLabel: translate("Label.Reason"),
          fieldValue: (violationReasons?.translatedReasons ?? []).join(", ") || translate("Label.AbuseType.Other")
        }
      ),
      /* @__PURE__ */ jsx19(
        EvidenceField_default,
        {
          fieldLabel: translate("Label.ModeratorNote"),
          fieldValue: messageToUser || translate("Description.Violation", { startLink: "", endLink: "" }),
          preline: true
        }
      )
    ] })
  ] });
};
var WhatHappenedPageItemConfig = {
  getIsVisible: () => true,
  renderComponent: WhatHappenedPageItem,
  configName: "what-happened"
};
var WhatHappenedPageItemConfig_default = WhatHappenedPageItemConfig;

// src/pageItemConfigs/educationalConfigs/createPolicyEducationPageItemConfig.tsx
import { jsx as jsx20, jsxs as jsxs14 } from "react/jsx-runtime";
var BULLET = "\u2022";
var splitByNewlines = (text) => {
  return text.split("\n").map((line) => line.trim()).filter(Boolean);
};
var createPolicyEducationPageItemConfig = (content) => {
  const PolicyEducationPageItem = () => {
    const translate = useNotApprovedTranslate();
    const translatedDescription = translate(content.description);
    const descriptionLines = splitByNewlines(translatedDescription);
    const translatedBullets = content.descriptionBullets ? translate(content.descriptionBullets) : void 0;
    const bulletLines = translatedBullets ? splitByNewlines(translatedBullets) : [];
    return /* @__PURE__ */ jsxs14("div", { className: "flex flex-col gap-medium", "data-testid": content.policyKey, children: [
      /* @__PURE__ */ jsx20("span", { className: "text-heading-medium", children: translate(content.title) }),
      content.subtitle && /* @__PURE__ */ jsx20("span", { className: "text-title-large", children: translate(content.subtitle) }),
      /* @__PURE__ */ jsxs14("div", { className: "flex flex-col gap-small", children: [
        descriptionLines.length > 0 && /* @__PURE__ */ jsx20("div", { className: "flex flex-col gap-medium", children: descriptionLines.map((line) => /* @__PURE__ */ jsx20("p", { className: "text-body-large", children: line }, line)) }),
        bulletLines.length > 0 && /* @__PURE__ */ jsx20("div", { className: "flex flex-col gap-xsmall padding-left-small", children: bulletLines.map((line) => /* @__PURE__ */ jsxs14("div", { className: "flex flex-row gap-small", children: [
          /* @__PURE__ */ jsx20("p", { className: "text-body-large", children: BULLET }),
          /* @__PURE__ */ jsx20("p", { className: "text-body-large", children: line })
        ] }, line)) })
      ] })
    ] });
  };
  return {
    getIsVisible: () => true,
    renderComponent: PolicyEducationPageItem,
    configName: content.policyKey
  };
};
var createPolicyEducationPageItemConfig_default = createPolicyEducationPageItemConfig;

// src/components/cta/ContinueButtonCta.tsx
import { Button as Button5 } from "@rbx/foundation-ui";

// src/components/cta/ReactivationCoachmark.tsx
import { useState as useState5 } from "react";
import {
  EducationalTooltip,
  EducationalTooltipTrigger,
  EducationalTooltipContent,
  EducationalTooltipBody,
  EducationalTooltipTitle,
  EducationalTooltipDescription
} from "@rbx/foundation-ui";

// src/utils/canReactivate.ts
function canReactivate(punishmentTypeDescription, endDate, verificationCategory) {
  if (verificationCategory === VERIFICATION_CATEGORIES.Email || verificationCategory === VERIFICATION_CATEGORIES.VPC) {
    return true;
  }
  const endDateTimeObject = new Date(endDate);
  const isPastEndDate = endDateTimeObject < /* @__PURE__ */ new Date();
  return punishmentTypeDescription === "Warn" /* Warn */ || punishmentTypeDescription !== "Delete" /* Delete */ && isPastEndDate;
}

// src/utils/hasAnyCommutation.ts
function hasAnyCommutation(commutationEligibility) {
  if (!commutationEligibility) return false;
  return Object.values(commutationEligibility).some(Boolean);
}

// src/utils/useIsMediumScreen.ts
import { useEffect, useState as useState4 } from "react";
var MEDIUM_BREAKPOINT_QUERY = "(min-width: 601px)";
function useIsMediumScreen() {
  const [matches, setMatches] = useState4(() => window.matchMedia(MEDIUM_BREAKPOINT_QUERY).matches);
  useEffect(() => {
    const mediaQueryList = window.matchMedia(MEDIUM_BREAKPOINT_QUERY);
    const onChange = () => {
      setMatches(mediaQueryList.matches);
    };
    mediaQueryList.addEventListener("change", onChange);
    return () => {
      mediaQueryList.removeEventListener("change", onChange);
    };
  }, []);
  return matches;
}

// src/components/cta/ReactivationCoachmark.tsx
import { jsx as jsx21, jsxs as jsxs15 } from "react/jsx-runtime";
var ReactivationCoachmark = ({ children }) => {
  const translate = useNotApprovedTranslate();
  const { commutationEligibility, ixpData, punishmentData } = useNotApprovedPagePunishment();
  const { hasNavigatedForward } = usePageNavigation();
  const isMediumScreen = useIsMediumScreen();
  const [isOpen, setIsOpen] = useState5(true);
  const isEligible = punishmentData && ixpData?.enableReactivationNotice && canReactivate(
    punishmentData.punishmentTypeDescription,
    punishmentData.endDate,
    punishmentData.verificationCategory
  ) && !punishmentData.verificationCategory && !hasAnyCommutation(commutationEligibility) && !hasNavigatedForward;
  return /* @__PURE__ */ jsxs15(EducationalTooltip, { open: isEligible && isOpen, onOpenChange: setIsOpen, children: [
    /* @__PURE__ */ jsx21(EducationalTooltipTrigger, { asChild: true, children }),
    /* @__PURE__ */ jsx21(
      EducationalTooltipContent,
      {
        position: isMediumScreen ? "top-end" : "top-center",
        hasCloseAffordance: true,
        onOpenAutoFocus: (event) => {
          event.preventDefault();
        },
        closeLabel: translate("Action.Close"),
        children: /* @__PURE__ */ jsxs15(EducationalTooltipBody, { children: [
          /* @__PURE__ */ jsx21(EducationalTooltipTitle, { children: translate("Heading.ReactivationNotice") }),
          /* @__PURE__ */ jsx21(EducationalTooltipDescription, { children: translate("Description.ReactivationNotice") })
        ] })
      }
    )
  ] });
};
var ReactivationCoachmark_default = ReactivationCoachmark;

// src/components/cta/ContinueButtonCta.tsx
import { jsx as jsx22 } from "react/jsx-runtime";
var ContinueButtonCta = ({ isDisabled = false }) => {
  const translate = useNotApprovedTranslate();
  const { goToNextPage } = usePageNavigation();
  return /* @__PURE__ */ jsx22(ReactivationCoachmark_default, { children: /* @__PURE__ */ jsx22(
    Button5,
    {
      variant: "Emphasis",
      size: "Medium",
      onClick: goToNextPage,
      "data-testid": "continue-button",
      isDisabled,
      children: translate("Action.Continue")
    }
  ) });
};
var ContinueButtonCta_default = ContinueButtonCta;

// src/components/cta/ProceedActionsCta.tsx
import { useState as useState7, useEffect as useEffect3, Fragment as Fragment2 } from "react";
import { Checkbox } from "@rbx/foundation-ui";
import { EventTypes as EventTypes8 } from "@rbx/not-approved-page-events";

// src/components/cta/NotApprovedProceedButton.tsx
import { Fragment, useState as useState6 } from "react";
import { useMutation } from "@tanstack/react-query";
import { Button as Button6 } from "@rbx/foundation-ui";
import { EventTypes as EventTypes7 } from "@rbx/not-approved-page-events";

// src/services/reactivateAccount.ts
var reactivateAccount = (userModerationApiUrl, httpPost) => httpPost(`${userModerationApiUrl}/v1/not-approved/reactivate`);
var reactivateAccount_default = reactivateAccount;

// src/utils/types.ts
var ProceedAction = /* @__PURE__ */ ((ProceedAction2) => {
  ProceedAction2[ProceedAction2["Reactivate"] = 0] = "Reactivate";
  ProceedAction2[ProceedAction2["Paused"] = 1] = "Paused";
  ProceedAction2[ProceedAction2["VerifyEmail"] = 2] = "VerifyEmail";
  ProceedAction2[ProceedAction2["VerifyVPC"] = 3] = "VerifyVPC";
  return ProceedAction2;
})(ProceedAction || {});

// src/context/PageAnalyticsContext.tsx
import {
  createContext as createContext4,
  useContext as useContext4,
  useMemo as useMemo3,
  useCallback as useCallback3,
  useRef,
  useEffect as useEffect2
} from "react";
import { EventTypes as EventTypes6 } from "@rbx/not-approved-page-events";
import { jsx as jsx23 } from "react/jsx-runtime";
var PageAnalyticsContext = createContext4(void 0);
var PageAnalyticsProvider = ({ children }) => {
  const { currentPage, currentPageName, unmappedViolationKeys } = usePageNavigation();
  const sendEvent = useSendNotApprovedPageEvent();
  const pageEnterTimeRef = useRef(Date.now());
  const previousPageRef = useRef(currentPage);
  const previousPageNameRef = useRef(currentPageName);
  const isInitialMountRef = useRef(true);
  const getTimeOnPageMs = () => {
    return Date.now() - pageEnterTimeRef.current;
  };
  useEffect2(() => {
    unmappedViolationKeys.forEach((unmappedViolationKey) => {
      sendEvent(EventTypes6.UnmappedViolationKey, { unmappedViolationKey });
    });
  }, [unmappedViolationKeys]);
  useEffect2(() => {
    if (isInitialMountRef.current) {
      isInitialMountRef.current = false;
      return;
    }
    const timeOnPageMs = getTimeOnPageMs();
    const eventType = currentPage > previousPageRef.current ? EventTypes6.ContinueClicked : EventTypes6.BackClicked;
    sendEvent(eventType, {
      pageName: previousPageNameRef.current,
      timeOnPageMs
    });
    pageEnterTimeRef.current = Date.now();
    previousPageRef.current = currentPage;
    previousPageNameRef.current = currentPageName;
  }, [currentPage, currentPageName]);
  const sendPageEvent = useCallback3(
    (eventType, additionalProperties) => {
      const timeOnPageMs = getTimeOnPageMs();
      sendEvent(eventType, {
        ...additionalProperties,
        pageName: currentPageName,
        timeOnPageMs
      });
    },
    [currentPageName, sendEvent]
  );
  const contextValue = useMemo3(
    () => ({ sendPageEvent }),
    [sendPageEvent]
  );
  return /* @__PURE__ */ jsx23(PageAnalyticsContext.Provider, { value: contextValue, children });
};
var usePageAnalytics = () => {
  const context = useContext4(PageAnalyticsContext);
  if (!context) {
    throw new Error("usePageAnalytics must be used within PageAnalyticsProvider");
  }
  return context;
};

// src/components/ErrorAlert.tsx
import { Icon, IconButton } from "@rbx/foundation-ui";
import { jsx as jsx24, jsxs as jsxs16 } from "react/jsx-runtime";
var ErrorAlert = ({ onClose }) => {
  const translate = useNotApprovedTranslate();
  return /* @__PURE__ */ jsxs16(
    "div",
    {
      className: "width-full flex gap-large items-center justify-between radius-medium padding-medium stroke-system-alert stroke-standard maf-error-alert",
      style: { backgroundColor: "rgba(255, 0, 0, 0.1)" },
      "data-testid": "error-alert",
      children: [
        /* @__PURE__ */ jsxs16("div", { className: "flex gap-small items-center", children: [
          /* @__PURE__ */ jsx24(Icon, { name: "icon-filled-circle-x", className: "content-system-alert" }),
          /* @__PURE__ */ jsx24("span", { className: "text-body-medium", children: translate("Heading.Error") })
        ] }),
        /* @__PURE__ */ jsx24(
          IconButton,
          {
            icon: "icon-regular-x-small",
            ariaLabel: "Close",
            onClick: onClose,
            variant: "Utility",
            size: "Medium",
            className: "content-system-alert shrink-0",
            isCircular: true
          }
        )
      ]
    }
  );
};
var ErrorAlert_default = ErrorAlert;

// src/components/cta/NotApprovedProceedButton.tsx
import { jsx as jsx25, jsxs as jsxs17 } from "react/jsx-runtime";
var NotApprovedProceedButton = ({
  proceedAction,
  setIsDialogOpen,
  isAgreed,
  isDisabled = false
}) => {
  const translate = useNotApprovedTranslate();
  const { sendPageEvent } = usePageAnalytics();
  const { userModerationApiUrl, httpPost, onVerifyEmail, onVerifyParent, onAccountReactivated } = useNotApprovedUIConfig();
  const [isLoading, setIsLoading] = useState6(false);
  const reactivateMutation = useMutation({
    mutationFn: () => reactivateAccount_default(userModerationApiUrl, httpPost),
    onSuccess: async () => {
      await new Promise((resolve) => {
        setTimeout(resolve, REACTIVATION_CACHE_UPDATE_WAIT);
      });
      onAccountReactivated();
    },
    onMutate: () => {
      setIsLoading(true);
      sendPageEvent(EventTypes7.ReactivateClicked);
    },
    onError: (error) => {
      const message = error instanceof Error ? error.message : "Unknown error";
      console.error("reactivateMutation error", message);
      setIsLoading(false);
    },
    retry: 0
  });
  let buttonText;
  switch (proceedAction) {
    case 1 /* Paused */:
    case 0 /* Reactivate */: {
      buttonText = translate("Action.Continue");
      break;
    }
    case 2 /* VerifyEmail */: {
      buttonText = translate("Label.EmailVerificationButtonText");
      break;
    }
    case 3 /* VerifyVPC */: {
      buttonText = translate("Label.ParentVerificationButtonText");
      break;
    }
    case null: {
      console.warn(`proceedAction has impossible value ${proceedAction}`);
      buttonText = "";
      break;
    }
  }
  const handleOnClick = async () => {
    switch (proceedAction) {
      case 0 /* Reactivate */: {
        reactivateMutation.mutate();
        return;
      }
      case 2 /* VerifyEmail */: {
        setIsDialogOpen(false);
        sendPageEvent(EventTypes7.EmailVerificationClicked);
        await onVerifyEmail?.();
        return;
      }
      case 3 /* VerifyVPC */: {
        setIsDialogOpen(false);
        sendPageEvent(EventTypes7.ParentVerificationClicked);
        await onVerifyParent?.({
          featureName: ACCESS_MANAGEMENT_UPSELL_CAN_LIFT_PUNISHMENT_FEATURE_NAME,
          ampRecourseData: { punishmentType: "Chargeback" /* Chargeback */ },
          isAsyncCall: true,
          usePrologue: false
        });
        setIsDialogOpen(true);
        return;
      }
      case 1 /* Paused */: {
        console.warn("cannot reactivate suspended account yet");
        return;
      }
      case null: {
        console.warn(`proceedAction has impossible value ${proceedAction}`);
      }
    }
  };
  return /* @__PURE__ */ jsxs17(Fragment, { children: [
    reactivateMutation.isError && /* @__PURE__ */ jsx25(
      ErrorAlert_default,
      {
        onClose: () => {
          reactivateMutation.reset();
        }
      }
    ),
    /* @__PURE__ */ jsx25(
      Button6,
      {
        onClick: () => {
          handleOnClick().catch((error) => {
            const message = error instanceof Error ? error.message : "unknown";
            console.warn("NotApprovedProceedButton onClick error", message);
          });
        },
        isDisabled: isDisabled || isLoading || !isAgreed,
        isLoading,
        variant: "Emphasis",
        size: "Medium",
        "data-testid": "proceed-button",
        children: buttonText
      }
    )
  ] });
};
var NotApprovedProceedButton_default = NotApprovedProceedButton;

// src/utils/determineProceedAction.ts
function determineProceedAction(punishmentData) {
  const { punishmentTypeDescription, endDate, verificationCategory } = punishmentData;
  switch (verificationCategory) {
    case VERIFICATION_CATEGORIES.Email:
      return 2 /* VerifyEmail */;
    case VERIFICATION_CATEGORIES.VPC:
      return 3 /* VerifyVPC */;
    default:
      return canReactivate(punishmentTypeDescription, endDate, verificationCategory) ? 0 /* Reactivate */ : 1 /* Paused */;
  }
}
var determineProceedAction_default = determineProceedAction;

// src/components/cta/SuspensionDurationAlert.tsx
import { Icon as Icon2 } from "@rbx/foundation-ui";

// src/utils/getTimeLeftInSuspension.ts
var MS_PER_MINUTE = 1e3 * 60;
var MS_PER_HOUR = MS_PER_MINUTE * 60;
var MS_PER_DAY = MS_PER_HOUR * 24;
var getTimeLeftInSuspension = (endDateString, translate) => {
  const endDate = new Date(endDateString);
  const now = Date.now();
  if (Number.isNaN(endDate.getTime())) {
    return translate("Label.Hours", {
      hours: "00",
      minutes: "00"
    });
  }
  let diffMs = endDate.getTime() - now;
  if (diffMs < 0) {
    diffMs = 0;
  }
  if (diffMs > MS_PER_DAY) {
    const daysLeft = Math.ceil(diffMs / MS_PER_DAY);
    return translate("Label.Days", { number: daysLeft });
  }
  const hoursLeft = Math.floor(diffMs / MS_PER_HOUR);
  let minutesLeft = Math.floor(diffMs % MS_PER_HOUR / MS_PER_MINUTE);
  if (diffMs > 0 && diffMs < MS_PER_MINUTE) {
    minutesLeft = 1;
  }
  return translate("Label.Hours", {
    hours: hoursLeft.toString().padStart(2, "0"),
    minutes: minutesLeft.toString().padStart(2, "0")
  });
};
var getTimeLeftInSuspension_default = getTimeLeftInSuspension;

// src/components/cta/SuspensionDurationAlert.tsx
import { jsx as jsx26, jsxs as jsxs18 } from "react/jsx-runtime";
var SuspensionDurationAlert = ({ endDate }) => {
  const translate = useNotApprovedTranslate();
  const endDateString = getTimeLeftInSuspension_default(endDate, translate);
  return /* @__PURE__ */ jsxs18(
    "div",
    {
      "data-testid": "suspension-duration-alert",
      className: "flex flex-wrap gap-small bg-shift-100 padding-x-medium padding-y-small stroke-standard stroke-default items-center radius-medium",
      children: [
        /* @__PURE__ */ jsx26(Icon2, { name: "icon-filled-triangle-exclamation", className: "content-system-warning" }),
        /* @__PURE__ */ jsx26("span", { className: "text-title-medium", children: translate("Label.Suspension") }),
        /* @__PURE__ */ jsx26("p", { className: "text-body-medium", children: endDateString })
      ]
    }
  );
};
var SuspensionDurationAlert_default = SuspensionDurationAlert;

// src/utils/isBanType.ts
function isBanType(punishmentTypeDescription) {
  const isValidPunishmentType = Object.values(PUNISHMENT_TYPE).includes(punishmentTypeDescription);
  return isValidPunishmentType && punishmentTypeDescription.startsWith("Ban");
}
var isBanType_default = isBanType;

// src/components/cta/ProceedActionsCta.tsx
import { jsx as jsx27, jsxs as jsxs19 } from "react/jsx-runtime";
var ONE_MINUTE_MS = 60 * 1e3;
var ProceedActionsCta = ({ punishmentData, setIsDialogOpen }) => {
  const translate = useNotApprovedTranslate();
  const sendEvent = useSendNotApprovedPageEvent();
  const { endDate, punishmentTypeDescription } = punishmentData;
  const proceedAction = determineProceedAction_default(punishmentData);
  const [isAgreed, setIsAgreed] = useState7(false);
  const [toggleRefresh, setToggleRefresh] = useState7(false);
  useEffect3(() => {
    let timer;
    if (isBanType_default(punishmentTypeDescription)) {
      timer = setTimeout(() => {
        setToggleRefresh((prev) => !prev);
      }, ONE_MINUTE_MS);
    }
    return () => {
      clearTimeout(timer);
    };
  }, [toggleRefresh, punishmentTypeDescription]);
  const onCheckboxChange = (isChecked) => {
    setIsAgreed(isChecked === true);
    sendEvent(EventTypes8.CheckboxChecked);
  };
  const isPaused = proceedAction === 1 /* Paused */;
  return isPaused ? /* @__PURE__ */ jsxs19("div", { className: "flex flex-col gap-large medium:flex-row", children: [
    /* @__PURE__ */ jsx27(SuspensionDurationAlert_default, { endDate }),
    /* @__PURE__ */ jsx27(
      NotApprovedProceedButton_default,
      {
        proceedAction,
        setIsDialogOpen,
        isAgreed,
        isDisabled: true
      }
    )
  ] }) : /* @__PURE__ */ jsxs19(Fragment2, { children: [
    /* @__PURE__ */ jsx27(
      Checkbox,
      {
        label: translate("Label.RuleAcknowledgment"),
        placement: "Start",
        size: "Small",
        isChecked: isAgreed,
        onCheckedChange: onCheckboxChange,
        className: "self-start",
        "data-testid": "rule-confirmation-checkbox"
      }
    ),
    /* @__PURE__ */ jsx27(
      NotApprovedProceedButton_default,
      {
        proceedAction,
        setIsDialogOpen,
        isAgreed,
        isDisabled: !isAgreed
      }
    )
  ] });
};
var ProceedActionsCta_default = ProceedActionsCta;

// src/components/cta/ReportMistakeButtonCta.tsx
import { Button as Button7 } from "@rbx/foundation-ui";
import { EventTypes as EventTypes9 } from "@rbx/not-approved-page-events";
import { jsx as jsx28 } from "react/jsx-runtime";
var ReportMistakeButtonCta = () => {
  const translate = useNotApprovedTranslate();
  const sendEvent = useSendNotApprovedPageEvent();
  const { websiteUrl } = useNotApprovedUIConfig();
  const handleReportMistake = () => {
    sendEvent(EventTypes9.AppealsPortalClicked);
    window.open(`${websiteUrl}/report-appeals?t_source=nap-web`, "_blank");
  };
  return /* @__PURE__ */ jsx28(
    Button7,
    {
      variant: "SoftEmphasis",
      size: "Medium",
      onClick: handleReportMistake,
      "data-testid": "appeals-button",
      children: translate("Action.ReportMistake")
    }
  );
};
var ReportMistakeButtonCta_default = ReportMistakeButtonCta;

// src/components/cta/SecondChanceActionsCta.tsx
import { Fragment as Fragment3, useState as useState8 } from "react";
import { useMutation as useMutation2 } from "@tanstack/react-query";
import { Button as Button8 } from "@rbx/foundation-ui";
import { EventTypes as EventTypes10 } from "@rbx/not-approved-page-events";

// src/services/reactivateAccountByCommutation.ts
var reactivateAccountByCommutation = async (apiGatewayUrl, httpPost) => {
  return httpPost(`${apiGatewayUrl}/moderation-appeal-service/v2/consequence-commutation`, {
    type: "EDUCATIONAL_PASS" /* Educational */
  });
};
var reactivateAccountByCommutation_default = reactivateAccountByCommutation;

// src/components/cta/SecondChanceActionsCta.tsx
import { jsx as jsx29, jsxs as jsxs20 } from "react/jsx-runtime";
var SecondChanceActionsCta = () => {
  const translate = useNotApprovedTranslate();
  const { sendPageEvent } = usePageAnalytics();
  const { httpPost, apiGatewayUrl, websiteUrl, onAccountReactivated } = useNotApprovedUIConfig();
  const [isReactivateLoading, setIsReactivateLoading] = useState8(false);
  const reactivateMutation = useMutation2({
    mutationFn: () => reactivateAccountByCommutation_default(apiGatewayUrl, httpPost),
    onMutate: () => {
      setIsReactivateLoading(true);
      sendPageEvent(EventTypes10.SecondChanceReactivateClicked);
    },
    onSuccess: async () => {
      await new Promise((resolve) => {
        setTimeout(resolve, REACTIVATION_CACHE_UPDATE_WAIT);
      });
      onAccountReactivated();
    },
    onError: (error) => {
      const message = error instanceof Error ? error.message : "Unknown error";
      console.error("reactivateMutation error", message);
      setIsReactivateLoading(false);
    },
    retry: 0
    // pin to prevent double-firing onMutate if QueryClient defaults change
  });
  const handleSendAppeal = () => {
    sendPageEvent(EventTypes10.AppealsPortalClicked);
    window.open(`${websiteUrl}/report-appeals?t_source=nap-web`, "_blank");
  };
  return /* @__PURE__ */ jsxs20(Fragment3, { children: [
    reactivateMutation.isError && /* @__PURE__ */ jsx29(
      ErrorAlert_default,
      {
        onClose: () => {
          reactivateMutation.reset();
        }
      }
    ),
    /* @__PURE__ */ jsxs20("div", { className: "flex flex-col gap-small medium:flex-row-reverse", children: [
      /* @__PURE__ */ jsx29(
        Button8,
        {
          onClick: () => {
            reactivateMutation.mutate();
          },
          variant: "Emphasis",
          size: "Medium",
          className: "min-width-2600",
          "data-testid": "second-chance-reactivate-button",
          isLoading: isReactivateLoading,
          isDisabled: isReactivateLoading,
          children: translate("Action.OK")
        }
      ),
      /* @__PURE__ */ jsx29(
        Button8,
        {
          onClick: handleSendAppeal,
          variant: "Standard",
          size: "Medium",
          "data-testid": "second-chance-send-appeal-button",
          children: translate("Action.SendAppeal")
        }
      )
    ] })
  ] });
};
var SecondChanceActionsCta_default = SecondChanceActionsCta;

// src/components/cta/UnderstandContinueCta.tsx
import { Fragment as Fragment4, useState as useState9 } from "react";
import { Checkbox as Checkbox2 } from "@rbx/foundation-ui";
import { jsx as jsx30, jsxs as jsxs21 } from "react/jsx-runtime";
var UnderstandContinueCta = () => {
  const translate = useNotApprovedTranslate();
  const [isChecked, setIsChecked] = useState9(false);
  const onCheckboxChange = (checked) => {
    setIsChecked(checked === true);
  };
  return /* @__PURE__ */ jsxs21(Fragment4, { children: [
    /* @__PURE__ */ jsx30(
      Checkbox2,
      {
        label: translate("Label.RuleAcknowledgment"),
        placement: "Start",
        size: "Small",
        isChecked,
        onCheckedChange: onCheckboxChange,
        className: "self-start",
        "data-testid": "understand-continue-checkbox"
      }
    ),
    /* @__PURE__ */ jsx30(ContinueButtonCta_default, { isDisabled: !isChecked })
  ] });
};
var UnderstandContinueCta_default = UnderstandContinueCta;

// src/pageItemConfigs/generatePages.ts
var getEducationalPages = (violationTypeKeys, isEducationalPassEligible) => {
  const educationalPages = [];
  const unmappedViolationKeys = /* @__PURE__ */ new Set();
  const addedPolicyKeys = /* @__PURE__ */ new Set();
  violationTypeKeys.forEach((violationTypeKey) => {
    const policyEducationConfig = policyEducationContentRegistry_default[violationTypeKey];
    if (!policyEducationConfig) {
      unmappedViolationKeys.add(violationTypeKey);
      return;
    }
    if (addedPolicyKeys.has(policyEducationConfig.policyKey)) return;
    addedPolicyKeys.add(policyEducationConfig.policyKey);
    educationalPages.push({
      pageName: `policy-rule-${policyEducationConfig.policyKey}`,
      pageItems: [
        createPolicyEducationPageItemConfig_default({
          title: policyEducationConfig.ruleTitle,
          subtitle: policyEducationConfig.ruleSubtitle,
          description: policyEducationConfig.ruleDescription,
          descriptionBullets: policyEducationConfig.ruleDescriptionBullets,
          policyKey: policyEducationConfig.policyKey
        })
      ],
      CtaComponent: isEducationalPassEligible ? UnderstandContinueCta_default : ContinueButtonCta_default
    });
    educationalPages.push({
      pageName: `policy-importance-${policyEducationConfig.policyKey}`,
      pageItems: [
        createPolicyEducationPageItemConfig_default({
          title: policyEducationConfig.importanceTitle,
          description: policyEducationConfig.importanceDescription,
          policyKey: policyEducationConfig.policyKey
        })
      ],
      CtaComponent: isEducationalPassEligible ? UnderstandContinueCta_default : ContinueButtonCta_default
    });
  });
  return { educationalPages, unmappedViolationKeys: [...unmappedViolationKeys] };
};
var generatePages = (punishmentData, violationTypeKeys, commutationEligibility) => {
  const pages = [];
  const isDeletePunishment = punishmentData.punishmentTypeDescription === "Delete" /* Delete */;
  const isEducationalPassEligible = commutationEligibility?.educational_pass_eligible ?? false;
  pages.push({
    pageName: "intro" /* Intro */,
    pageItems: [
      PunishmentDescriptionPageItemConfig_default,
      WhatHappenedPageItemConfig_default,
      ReviewedEvidencePageItemConfig_default
    ],
    CtaComponent: isDeletePunishment ? ReportMistakeButtonCta_default : ContinueButtonCta_default
  });
  if (isDeletePunishment) {
    return { pages, unmappedViolationKeys: [] };
  }
  if (isEducationalPassEligible) {
    pages.push({
      pageName: "second-chance-intro" /* SecondChanceIntro */,
      pageItems: [SecondChanceIntroPageItemConfig_default, ReportMistakePageItemConfig_default],
      CtaComponent: ContinueButtonCta_default
    });
  }
  const { educationalPages, unmappedViolationKeys } = getEducationalPages(
    violationTypeKeys,
    isEducationalPassEligible
  );
  pages.push(...educationalPages);
  pages.push({
    pageName: "resolution" /* Resolution */,
    pageItems: [
      PreventionStepsPageItemConfig_default,
      ChargebackStepsPageItemConfig_default,
      ReportMistakePageItemConfig_default
    ],
    CtaComponent: isEducationalPassEligible ? UnderstandContinueCta_default : ProceedActionsCta_default
  });
  if (isEducationalPassEligible) {
    pages.push({
      pageName: "second-chance-conclusion" /* SecondChanceConclusion */,
      pageItems: [SecondChanceConclusionPageItemConfig_default],
      CtaComponent: SecondChanceActionsCta_default
    });
  }
  return { pages, unmappedViolationKeys };
};

// src/components/header/NotApprovedDialogHeader.tsx
import { Icon as Icon3 } from "@rbx/foundation-ui";

// src/components/header/HeaderPopoverMenu.tsx
import { useState as useState10 } from "react";
import { useMutation as useMutation3 } from "@tanstack/react-query";
import {
  Popover,
  Menu,
  MenuItem,
  PopoverContent,
  PopoverTrigger,
  ProgressCircle as ProgressCircle2,
  IconButton as IconButton2
} from "@rbx/foundation-ui";
import { EventTypes as EventTypes11 } from "@rbx/not-approved-page-events";
import { jsx as jsx31, jsxs as jsxs22 } from "react/jsx-runtime";
var HeaderPopoverMenu = () => {
  const translate = useNotApprovedTranslate();
  const { onLogout } = useNotApprovedUIConfig();
  const sendEvent = useSendNotApprovedPageEvent();
  const [isLogoutLoading, setIsLogoutLoading] = useState10(false);
  const logoutMutation = useMutation3({
    mutationFn: async () => {
      await onLogout();
    },
    onMutate: () => {
      setIsLogoutLoading(true);
      sendEvent(EventTypes11.LogoutClicked);
    },
    onError: (error) => {
      const message = error instanceof Error ? error.message : "Unknown error";
      sendEvent(EventTypes11.Error, {
        additionalInfo: `headerPopoverLogout: Error logging out - ${message}`
      });
      setIsLogoutLoading(false);
    },
    retry: 0
    // pin to prevent double-firing onMutate if QueryClient defaults change
  });
  return /* @__PURE__ */ jsxs22(Popover, { children: [
    /* @__PURE__ */ jsx31(PopoverTrigger, { asChild: true, children: /* @__PURE__ */ jsx31(
      IconButton2,
      {
        icon: "icon-filled-three-dots-vertical",
        ariaLabel: translate("Label.OpenMenu"),
        variant: "Utility",
        size: "Medium"
      }
    ) }),
    /* @__PURE__ */ jsx31(PopoverContent, { side: "bottom", align: "end", ariaLabel: translate("Label.MenuContent"), children: /* @__PURE__ */ jsx31(Menu, { size: "Medium", children: /* @__PURE__ */ jsx31(
      MenuItem,
      {
        value: "one",
        title: translate("Action.LogOut"),
        onSelect: () => {
          logoutMutation.mutate();
        },
        disabled: isLogoutLoading,
        trailing: isLogoutLoading ? /* @__PURE__ */ jsx31(
          ProgressCircle2,
          {
            size: "Small",
            ariaLabel: translate("Label.LogoutProgress"),
            variant: "Indeterminate"
          }
        ) : void 0
      }
    ) }) })
  ] });
};
var HeaderPopoverMenu_default = HeaderPopoverMenu;

// src/components/header/HeaderProgressBar.tsx
import { ProgressBar } from "@rbx/foundation-ui";
import { jsx as jsx32 } from "react/jsx-runtime";
var HeaderProgressBar = () => {
  const translate = useNotApprovedTranslate();
  const { getProgress } = usePageNavigation();
  const progress = getProgress();
  return /* @__PURE__ */ jsx32(
    ProgressBar,
    {
      value: progress,
      ariaLabel: translate("Label.PageProgress"),
      style: {
        visibility: progress === 0 ? "hidden" : "visible"
      }
    }
  );
};
var HeaderProgressBar_default = HeaderProgressBar;

// src/components/header/BackButton.tsx
import { IconButton as IconButton3 } from "@rbx/foundation-ui";
import { jsx as jsx33 } from "react/jsx-runtime";
var BackButton = () => {
  const translate = useNotApprovedTranslate();
  const { goToPreviousPage } = usePageNavigation();
  return /* @__PURE__ */ jsx33(
    IconButton3,
    {
      size: "Medium",
      variant: "Utility",
      icon: "icon-filled-chevron-large-left",
      onClick: goToPreviousPage,
      "data-testid": "back-button",
      className: "margin-left-[-12px]",
      ariaLabel: translate("Action.Back")
    }
  );
};
var BackButton_default = BackButton;

// src/components/header/NotApprovedDialogHeader.tsx
import { jsx as jsx34, jsxs as jsxs23 } from "react/jsx-runtime";
var NotApprovedDialogHeader = ({ punishmentData }) => {
  const { isFirstPage, totalPages } = usePageNavigation();
  const translate = useNotApprovedTranslate();
  const { punishmentTypeDescription, verificationCategory } = punishmentData;
  const punishmentTypeLabel = verificationCategory ? translate("Heading.Suspended") : translate(PUNISHMENT_TYPE_TO_STRING_KEY[punishmentTypeDescription] ?? "") || translate("Heading.Suspended");
  const iconColor = punishmentTypeDescription === "Warn" /* Warn */ && !verificationCategory ? "content-system-warning" : "content-system-alert";
  const iconName = punishmentTypeDescription === "Delete" /* Delete */ ? "icon-regular-circle-slash" : "icon-regular-triangle-exclamation";
  return /* @__PURE__ */ jsx34("div", { "data-testid": "not-approved-dialog-header", children: isFirstPage ? /* @__PURE__ */ jsxs23("div", { className: "flex justify-between items-center gap-medium", children: [
    /* @__PURE__ */ jsxs23("div", { className: "flex gap-small items-center padding-y-medium", children: [
      /* @__PURE__ */ jsx34(Icon3, { name: iconName, size: "Large", className: iconColor }),
      /* @__PURE__ */ jsx34("span", { className: "text-heading-small", children: punishmentTypeLabel })
    ] }),
    /* @__PURE__ */ jsx34(HeaderPopoverMenu_default, {})
  ] }) : /* @__PURE__ */ jsxs23("div", { className: "flex flex-col gap-large items-start", children: [
    /* @__PURE__ */ jsx34(BackButton_default, {}),
    totalPages > 2 && /* @__PURE__ */ jsx34(HeaderProgressBar_default, {})
  ] }) });
};
var NotApprovedDialogHeader_default = NotApprovedDialogHeader;

// src/components/cta/NotApprovedCtaButtons.tsx
import { jsx as jsx35 } from "react/jsx-runtime";
var NotApprovedCtaButtons = ({ punishmentData, setIsDialogOpen }) => {
  const { CurrentCtaComponent, currentPageName } = usePageNavigation();
  return CurrentCtaComponent ? /* @__PURE__ */ jsx35("div", { className: "flex flex-col gap-large justify-between medium:items-end", children: /* @__PURE__ */ jsx35(
    CurrentCtaComponent,
    {
      punishmentData,
      setIsDialogOpen
    },
    currentPageName
  ) }) : null;
};
var NotApprovedCtaButtons_default = NotApprovedCtaButtons;

// src/components/LoadingSkeleton.tsx
import { DialogBody, DialogTitle } from "@rbx/foundation-ui";
import { jsx as jsx36, jsxs as jsxs24 } from "react/jsx-runtime";
var LoadingSkeleton = () => {
  return /* @__PURE__ */ jsxs24(DialogBody, { className: "flex flex-col gap-xlarge height-[85vh] max-height-[800px]", children: [
    /* @__PURE__ */ jsxs24(DialogTitle, { className: "shrink-0 flex flex-row justify-between items-center gap-medium", children: [
      /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[45px] width-full max-width-[300px] radius-medium" }),
      /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[35px] width-[35px] radius-medium shrink-0" })
    ] }),
    /* @__PURE__ */ jsxs24(
      "div",
      {
        className: "grow-1 flex flex-col gap-large scroll-y",
        "data-testid": "loading-skeleton-content",
        children: [
          /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[20px] width-full max-width-[450px] radius-medium" }),
          /* @__PURE__ */ jsxs24("div", { className: "flex flex-col gap-medium", children: [
            /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[30px] width-[125px] radius-medium" }),
            /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[120px] width-full radius-medium" })
          ] }),
          /* @__PURE__ */ jsxs24("div", { className: "flex flex-col gap-medium", children: [
            /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[30px] width-[200px] radius-medium" }),
            /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[160px] width-full radius-medium" })
          ] })
        ]
      }
    ),
    /* @__PURE__ */ jsx36("div", { className: "shrink-0 flex width-full justify-end medium:flex-row", children: /* @__PURE__ */ jsx36("div", { className: "bg-shift-100 height-[40px] width-full medium:width-[90px] radius-medium" }) })
  ] });
};
var LoadingSkeleton_default = LoadingSkeleton;

// src/components/ErrorMessage.tsx
import { DialogBody as DialogBody2, DialogTitle as DialogTitle2, Icon as Icon4 } from "@rbx/foundation-ui";
import { jsx as jsx37, jsxs as jsxs25 } from "react/jsx-runtime";
var ErrorMessage = ({ error }) => {
  const translate = useNotApprovedTranslate();
  if (error) {
    console.error(error);
  }
  return /* @__PURE__ */ jsxs25(DialogBody2, { className: "flex flex-col gap-large", children: [
    /* @__PURE__ */ jsx37("div", { className: "shrink-0 flex justify-end items-center", children: /* @__PURE__ */ jsx37(HeaderPopoverMenu_default, {}) }),
    /* @__PURE__ */ jsxs25("div", { className: "grow-1 flex flex-col items-center justify-center gap-medium padding-xxlarge radius-medium bg-surface-100 width-full min-height-[225px]", children: [
      /* @__PURE__ */ jsx37(
        Icon4,
        {
          name: "icon-filled-triangle-exclamation",
          size: "XLarge",
          className: "content-system-alert"
        }
      ),
      /* @__PURE__ */ jsx37(DialogTitle2, { className: "text-heading-medium", children: translate("Heading.Error") })
    ] })
  ] });
};
var ErrorMessage_default = ErrorMessage;

// src/components/GenericFallbackDialog.tsx
import { useEffect as useEffect4, useState as useState11 } from "react";
import { useMutation as useMutation4 } from "@tanstack/react-query";
import {
  Button as Button9,
  Dialog,
  DialogBody as DialogBody3,
  DialogContent,
  DialogFooter,
  DialogTitle as DialogTitle3
} from "@rbx/foundation-ui";
import { EventTypes as EventTypes12 } from "@rbx/not-approved-page-events";
import { jsx as jsx38, jsxs as jsxs26 } from "react/jsx-runtime";
var GenericFallbackDialog = ({
  impressionEvent
}) => {
  const { translate, websiteUrl, onLogout } = useNotApprovedUIConfig();
  const sendEvent = useSendNotApprovedPageEvent();
  const translateWithLink = makeTranslateWithLink_default(translate, sendEvent);
  const [isLoggingOut, setIsLoggingOut] = useState11(false);
  useEffect4(() => {
    sendEvent(impressionEvent);
  }, [sendEvent, impressionEvent]);
  const logoutMutation = useMutation4({
    mutationFn: async () => {
      await onLogout();
    },
    onMutate: () => {
      setIsLoggingOut(true);
      sendEvent(EventTypes12.LogoutClicked);
    },
    onError: (error) => {
      const message = error instanceof Error ? error.message : "Unknown error";
      sendEvent(EventTypes12.Error, {
        additionalInfo: `${impressionEvent}: Error logging out - ${message}`
      });
      setIsLoggingOut(false);
    },
    retry: 0
  });
  const handleGoToRoblox = () => {
    window.open(websiteUrl, "_blank", "noopener,noreferrer");
  };
  return /* @__PURE__ */ jsx38(Dialog, { open: true, isModal: true, size: "Medium", hasCloseAffordance: false, children: /* @__PURE__ */ jsxs26(DialogContent, { className: "[&_p]:margin-none [&_h2]:margin-none", children: [
    /* @__PURE__ */ jsxs26(DialogBody3, { className: "gap-large flex flex-col", children: [
      /* @__PURE__ */ jsx38(DialogTitle3, { className: "text-heading-large margin-none", children: translate("Heading.AccountIssue") }),
      translateWithLink("Description.ResolveIssue", websiteUrl)
    ] }),
    /* @__PURE__ */ jsxs26(DialogFooter, { className: "flex justify-end gap-small flex-col-reverse medium:flex-row", children: [
      /* @__PURE__ */ jsx38(
        Button9,
        {
          variant: "Standard",
          size: "Medium",
          isLoading: isLoggingOut,
          isDisabled: isLoggingOut,
          onClick: () => {
            logoutMutation.mutate();
          },
          children: translate("Action.Logout")
        }
      ),
      /* @__PURE__ */ jsx38(Button9, { variant: "Emphasis", size: "Medium", onClick: handleGoToRoblox, children: translate("Action.GoToRoblox") })
    ] })
  ] }) });
};
var GenericFallbackDialog_default = GenericFallbackDialog;

// src/components/NotApprovedMainPage.tsx
import { jsx as jsx39, jsxs as jsxs27 } from "react/jsx-runtime";
var NotApprovedMainPage = () => {
  const {
    surfaceColor,
    renderSelfServiceDeactivated,
    shouldShowGenericFallback,
    onNoPunishmentData,
    ixp
  } = useNotApprovedUIConfig();
  const sendEvent = useSendNotApprovedPageEvent();
  const {
    punishmentData,
    violationReasons,
    isLoading,
    error,
    isReady,
    commutationEligibility,
    ixpData
  } = useNotApprovedPagePunishment();
  const [isDialogOpen, setIsDialogOpen] = useState12(true);
  const isEnrolledInReactivationNotice = ixpData && "enableReactivationNotice" in ixpData;
  const untranslatedReasons = violationReasons?.untranslatedReasons;
  const { pages, unmappedViolationKeys } = useMemo4(() => {
    return !punishmentData ? { pages: [], unmappedViolationKeys: [] } : generatePages(punishmentData, untranslatedReasons ?? [], commutationEligibility);
  }, [punishmentData, untranslatedReasons, commutationEligibility]);
  const hasEducationalPages = useMemo4(
    () => (untranslatedReasons ?? []).some((key) => key in policyEducationContentRegistry_default),
    [untranslatedReasons]
  );
  useEffect5(() => {
    if (punishmentData) {
      const eventType = punishmentData.context?.SelfServiceDeactivated ? EventTypes13.AccountReactivationPageRendered : EventTypes13.PageRendered;
      sendEvent(eventType, {
        interventionId: punishmentData.interventionId,
        punishedUserId: punishmentData.punishedUserId,
        isReactivationEligible: canReactivate(
          punishmentData.punishmentTypeDescription,
          punishmentData.endDate,
          punishmentData.verificationCategory
        ),
        verificationCategory: punishmentData.verificationCategory
      });
    }
  }, [punishmentData]);
  useEffect5(() => {
    if (isEnrolledInReactivationNotice && punishmentData && canReactivate(
      punishmentData.punishmentTypeDescription,
      punishmentData.endDate,
      punishmentData.verificationCategory
    ) && !punishmentData.verificationCategory && !hasAnyCommutation(commutationEligibility)) {
      ixp?.logExposure(IXP_LAYER_NAME);
    }
  }, [isEnrolledInReactivationNotice, punishmentData, commutationEligibility, ixp]);
  const hasNoPunishmentData = !isLoading && !error && !punishmentData;
  useEffect5(() => {
    if (hasNoPunishmentData) {
      onNoPunishmentData?.();
    }
  }, [hasNoPunishmentData]);
  if (hasNoPunishmentData) {
    return null;
  }
  if (isReady && punishmentData?.context?.SelfServiceDeactivated) {
    return renderSelfServiceDeactivated ? /* @__PURE__ */ jsx39(Fragment5, { children: renderSelfServiceDeactivated() }) : /* @__PURE__ */ jsx39(GenericFallbackDialog_default, { impressionEvent: EventTypes13.AccountReactivationRedirectRendered });
  }
  const fallbackImpressionEvent = isReady && punishmentData && shouldShowGenericFallback?.(punishmentData);
  if (fallbackImpressionEvent) {
    return /* @__PURE__ */ jsx39(GenericFallbackDialog_default, { impressionEvent: fallbackImpressionEvent });
  }
  return /* @__PURE__ */ jsx39(Dialog2, { open: isDialogOpen, size: "Large", hasCloseAffordance: false, isModal: true, children: /* @__PURE__ */ jsxs27(
    DialogContent2,
    {
      className: "width-full [&_p]:margin-none [&_h2]:margin-none",
      "aria-describedby": void 0,
      onOpenAutoFocus: (event) => {
        event.preventDefault();
      },
      children: [
        isLoading && /* @__PURE__ */ jsx39(LoadingSkeleton_default, {}),
        !isLoading && error && /* @__PURE__ */ jsx39(ErrorMessage_default, { error }),
        isReady && punishmentData && /* @__PURE__ */ jsx39(
          PageNavigationProvider,
          {
            pages,
            unmappedViolationKeys,
            hasEducationalPages,
            commutationEligibility,
            children: /* @__PURE__ */ jsx39(PageAnalyticsProvider, { children: /* @__PURE__ */ jsxs27(DialogBody4, { className: "flex flex-col gap-large height-[85vh] max-height-[800px]", children: [
              /* @__PURE__ */ jsx39(DialogTitle4, { className: "shrink-0", children: /* @__PURE__ */ jsx39(NotApprovedDialogHeader_default, { punishmentData }) }),
              /* @__PURE__ */ jsxs27("div", { className: "grow-1 scroll-y", children: [
                /* @__PURE__ */ jsx39(
                  PageItemsFromConfigs_default,
                  {
                    punishmentData,
                    commutationEligibility
                  }
                ),
                /* @__PURE__ */ jsx39(
                  "div",
                  {
                    style: {
                      position: "sticky",
                      bottom: -1,
                      left: 0,
                      height: "40px",
                      background: `linear-gradient(to bottom, transparent 0%, ${surfaceColor} 100%)`
                    }
                  }
                )
              ] }),
              /* @__PURE__ */ jsx39("div", { className: "shrink-0", children: /* @__PURE__ */ jsx39(
                NotApprovedCtaButtons_default,
                {
                  punishmentData,
                  setIsDialogOpen
                }
              ) })
            ] }) })
          }
        )
      ]
    }
  ) });
};
var NotApprovedMainPage_default = NotApprovedMainPage;

// src/containers/NotApprovedPageContainer.tsx
import { jsx as jsx40 } from "react/jsx-runtime";
var ONE_MIN_MS = 60 * 1e3;
var queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      refetchOnWindowFocus: false,
      retry: false,
      refetchOnMount: false,
      staleTime: ONE_MIN_MS
    }
  }
});
function NotApprovedPageContainer() {
  return /* @__PURE__ */ jsx40(QueryClientProvider, { client: queryClient, children: /* @__PURE__ */ jsx40(NotApprovedPagePunishmentProvider, { enableIxp: true, children: /* @__PURE__ */ jsx40(NotApprovedMainPage_default, {}) }) });
}
var NotApprovedPageContainer_default = NotApprovedPageContainer;

// src/translationConfig.ts
var translationConfig = [
  "Feature.Moderation",
  "Common.AlertsAndOptions",
  "CommonUI.Controls",
  "Feature.AppealsPortal",
  "Feature.NotApproved"
];
export {
  NotApprovedPageContainer_default as NotApprovedPageContainer,
  NotApprovedUIProvider,
  ProceedAction,
  sendNotApprovedPageEvent_default as sendNotApprovedPageEvent,
  translationConfig,
  useNotApprovedTranslate,
  useNotApprovedUIConfig,
  useSendNotApprovedPageEvent
};
