// @rbx/creator-hub-search/src/search/hooks/useStoreSearch.ts
import { useEffect, useRef, useState } from 'react';
import type { Locale } from '@rbx/intl';
// eslint-disable-next-line consistent-type-imports -- need value access for SearchDisplayCategory.Store
import { SearchDisplayCategory } from '../../clients/docSiteSearchType';
import { useToolboxSearchClient } from '../../clients/toolboxSearchClient';
import { useSearchConfig } from '../../contexts/SearchConfigContext';
import { ESearchInteraction } from '../../eventStream/enum/DocsSiteSearch';
import throwError from '../../utilities/error';
import { trackSearchPerformed } from '../searchEvents';
import type { TSearchListItem } from '../types/SearchListItem';
import { mapStoreResultsToListItems } from '../utils/storeSearchTransform';

const NUM_RESULTS_TO_GET = 14;

export interface UseStoreSearchOptions {
  query: string;
  isContentTypeFilterDefault: boolean;
  isDisplayFilterDefault: boolean;
  displayFilter: SearchDisplayCategory;
  searchSessionId: string;
  locale: Locale;
  translate: (key: string, params?: Record<string, string>) => string;
}

export interface UseStoreSearchReturn {
  storeResults: TSearchListItem[];
  isStoreLoading: boolean;
}

export function useStoreSearch({
  query,
  isContentTypeFilterDefault,
  isDisplayFilterDefault,
  displayFilter,
  searchSessionId,
  locale,
  translate,
}: UseStoreSearchOptions): UseStoreSearchReturn {
  const toolboxSearchClient = useToolboxSearchClient();
  const { eventLogger, currentProduct } = useSearchConfig();
  const [storeResults, setStoreResults] = useState<TSearchListItem[]>([]);
  const [isStoreLoading, setIsStoreLoading] = useState(false);
  // Per-session cache of mapped results keyed by query. Toggling the display
  // filter (All <-> Store, or drilling into Hub/Learn and back) re-runs this
  // effect with an unchanged query; without a cache that would re-issue the
  // multi-category fetch and re-fire the "search performed" analytics event,
  // double-counting Store impressions and deflating Store CTR. Scoped to the
  // hook instance (ref) so cached resultRefs keep this session's searchSessionId.
  const resultCacheRef = useRef<Map<string, TSearchListItem[]>>(new Map());

  useEffect(() => {
    if (query.length === 0) {
      setStoreResults([]);
      setIsStoreLoading(false);
      return undefined;
    }

    const shouldFetchStore =
      isContentTypeFilterDefault &&
      (isDisplayFilterDefault || displayFilter === SearchDisplayCategory.Store);

    if (!shouldFetchStore) {
      setStoreResults([]);
      setIsStoreLoading(false);
      return undefined;
    }

    // Already fetched this query in the current session: restore the cached
    // results without re-fetching or re-logging the performed event. (An empty
    // array is a valid cache hit, so check presence via `has`.)
    if (resultCacheRef.current.has(query)) {
      setStoreResults(resultCacheRef.current.get(query) ?? []);
      setIsStoreLoading(false);
      return undefined;
    }

    let invalidated = false;
    setIsStoreLoading(true);

    const fetchStore = async () => {
      try {
        const response = await toolboxSearchClient.search({
          query,
          limit: NUM_RESULTS_TO_GET,
        });
        if (invalidated) {
          return;
        }
        const items = mapStoreResultsToListItems(
          response.results,
          searchSessionId,
          locale,
          translate,
        );
        setStoreResults(items);
        setIsStoreLoading(false);
        // Don't cache (or log) error responses so a later trigger can retry.
        if (!response.isError) {
          resultCacheRef.current.set(query, items);
          if (items.length > 0) {
            trackSearchPerformed({
              eventLogger,
              category: 'Store',
              interaction: ESearchInteraction.Input,
              isRecommended: false,
              isSemantic: false,
              locale,
              query,
              currentProduct,
              searchResults: items.flatMap((item) => (item.resultRef ? [item.resultRef] : [])),
              searchSessionId,
            });
          }
        }
      } catch (e) {
        if (invalidated) {
          return;
        }
        throwError(
          `useStoreSearch: failed to fetch Store search results ${e instanceof Error ? e.message : String(e)}`,
        );
        setStoreResults([]);
        setIsStoreLoading(false);
      }
    };

    void fetchStore();

    return () => {
      invalidated = true;
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [query, isContentTypeFilterDefault, isDisplayFilterDefault, displayFilter]);

  return { storeResults, isStoreLoading };
}
