// @rbx/cueing/react/CueProvider.tsx
import React, { useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { emitCue } from '../core/emitCue';
import type { CueModalId } from '../core/ports';
import { shouldUserSeeCue } from '../core/shouldUserSeeCue';
import { CueLifecycle } from '../core/types';
import { CueContext, type UseCueResult } from './cueContext';
import { useCueingConfiguration } from './useCueingConfiguration';

export type CueProviderProps = {
  children: React.ReactNode;
  modalId: CueModalId;
};

const CueProvider = ({ children, modalId }: CueProviderProps) => {
  const configuration = useCueingConfiguration();
  const stepKeys = useMemo((): readonly string[] => {
    if (!configuration) {
      return [];
    }
    const config = configuration.registry[modalId];
    if (!config) {
      configuration.onError.log({ context: 'CueProvider', modalId, reason: 'unknown_modal_id' });
      return [];
    }
    return config.stepKeys ?? [];
  }, [configuration, modalId]);
  const tourMode = stepKeys.length > 1;

  const [eligible, setEligible] = useState<boolean>(false);
  const [eligibilityResolved, setEligibilityResolved] = useState<boolean>(false);
  const [stepIndex, setStepIndex] = useState<number>(0);
  const cueEndedRef = useRef(false);

  useEffect(() => {
    cueEndedRef.current = false;
    setStepIndex(0);
    setEligible(false);
    setEligibilityResolved(false);

    let cancelled = false;

    if (!configuration) {
      setEligibilityResolved(true);
      setEligible(false);
    } else {
      void shouldUserSeeCue(modalId).then((shouldSee) => {
        if (!cancelled) {
          setEligible(shouldSee);
          setEligibilityResolved(true);
        }
      });
    }

    return () => {
      cancelled = true;
    };
  }, [configuration, modalId]);

  const currentStepKey = stepKeys[stepIndex] ?? stepKeys[0] ?? '';

  const logStepShown = useCallback(
    (stepKey: string) => {
      emitCue({ lifecycle: CueLifecycle.Shown, modalId, stepKey }).catch(() => undefined);
    },
    [modalId],
  );

  const logSingleShown = useCallback(() => {
    emitCue({ lifecycle: CueLifecycle.Shown, modalId }).catch(() => undefined);
  }, [modalId]);

  // Shown analytics may fire more than once per step (e.g. tour back navigation). Terminal
  // lifecycles (closed / finished) and UI eligibility are deduped — see endCue.
  useEffect(() => {
    if (!eligibilityResolved || !eligible) {
      return;
    }
    if (tourMode && currentStepKey) {
      logStepShown(currentStepKey);
      return;
    }
    if (!tourMode) {
      logSingleShown();
    }
  }, [currentStepKey, eligibilityResolved, eligible, logSingleShown, logStepShown, tourMode]);

  const endCue = useCallback(
    (lifecycle: typeof CueLifecycle.Closed | typeof CueLifecycle.Finished, stepKey?: string) => {
      if (cueEndedRef.current) {
        return;
      }
      cueEndedRef.current = true;
      setEligible(false);
      emitCue({ lifecycle, modalId, stepKey }).catch(() => undefined);
    },
    [modalId],
  );

  const dismiss = useCallback(() => {
    endCue(CueLifecycle.Closed);
  }, [endCue]);

  const closeTour = useCallback(() => {
    const stepKey = stepKeys[stepIndex] ?? stepKeys[0] ?? '';
    endCue(CueLifecycle.Closed, stepKey || undefined);
  }, [endCue, stepIndex, stepKeys]);

  const goNext = useCallback(() => {
    if (cueEndedRef.current || !eligible) {
      return;
    }
    const isLastStep = stepIndex >= stepKeys.length - 1;
    if (isLastStep) {
      const stepKey = stepKeys[stepIndex] ?? stepKeys[0] ?? '';
      endCue(CueLifecycle.Finished, stepKey || undefined);
      return;
    }
    setStepIndex((index) => index + 1);
  }, [eligible, endCue, stepIndex, stepKeys]);

  const goBack = useCallback(() => {
    if (cueEndedRef.current || !eligible || stepIndex <= 0) {
      return;
    }
    setStepIndex((index) => index - 1);
  }, [eligible, stepIndex]);

  const contextValue = useMemo((): UseCueResult => {
    const resolvedEligible = eligibilityResolved ? eligible : false;

    if (tourMode) {
      return {
        closeTour,
        currentStepKey,
        eligible: resolvedEligible,
        goBack,
        goNext,
        modalId,
        mode: 'tour',
        stepIndex,
        totalSteps: stepKeys.length,
      };
    }

    return {
      dismiss,
      eligible: resolvedEligible,
      modalId,
      mode: 'single',
    };
  }, [
    closeTour,
    currentStepKey,
    dismiss,
    eligible,
    eligibilityResolved,
    goBack,
    goNext,
    modalId,
    stepIndex,
    stepKeys.length,
    tourMode,
  ]);

  return <CueContext.Provider value={contextValue}>{children}</CueContext.Provider>;
};

export default CueProvider;
