// @rbx/cueing/react/scrollCueAnchor.ts
const FOCUSABLE_SELECTOR =
  'button:not([disabled]), [href], input:not([disabled]), select:not([disabled]), textarea:not([disabled]), [tabindex]:not([tabindex="-1"])';

export type CueAnchorActivationOptions = {
  manageFocus?: boolean;
  scrollIntoView?: boolean;
};

export const DEFAULT_CUE_ANCHOR_ACTIVATION_OPTIONS: Required<CueAnchorActivationOptions> = {
  manageFocus: false,
  scrollIntoView: false,
};

export const focusCueAnchor = (element: HTMLElement): void => {
  const focusable = element.querySelector<HTMLElement>(FOCUSABLE_SELECTOR);
  (focusable ?? element).focus({ preventScroll: true });
};

export const scrollCueAnchorIntoView = (
  element: HTMLElement,
  options: CueAnchorActivationOptions = DEFAULT_CUE_ANCHOR_ACTIVATION_OPTIONS,
): void => {
  const { manageFocus, scrollIntoView } = { ...DEFAULT_CUE_ANCHOR_ACTIVATION_OPTIONS, ...options };
  if (scrollIntoView) {
    element.scrollIntoView({ behavior: 'smooth', block: 'nearest', inline: 'nearest' });
  }
  if (manageFocus) {
    focusCueAnchor(element);
  }
};
