// @rbx/analytics-ui/src/peripherals/ChartCardHeaderActions.tsx
import type { FC, ReactNode } from 'react';
import React, { memo, useCallback, useMemo, useState } from 'react';
import { Button, IconButton, Menu, MenuItem, MoreVertIcon, Tooltip } from '@rbx/ui';

type ChartCardHeaderActionBase = {
  readonly id: string;
  readonly label: string;
  readonly tooltip?: string;
  readonly disabled?: boolean;
  readonly testId?: string;
};

export type ChartCardHeaderButtonAction = ChartCardHeaderActionBase & {
  readonly kind: 'button';
  readonly onClick: () => void;
  readonly icon?: ReactNode;
  readonly renderButton?: (props: {
    readonly action: ChartCardHeaderButtonAction;
    readonly defaultButton: ReactNode;
  }) => ReactNode;
};

export type ChartCardHeaderLinkAction = ChartCardHeaderActionBase & {
  readonly kind: 'link';
  readonly href: string;
  readonly icon?: ReactNode;
  readonly renderLink?: (props: {
    readonly action: ChartCardHeaderLinkAction;
    readonly defaultLink: ReactNode;
  }) => ReactNode;
};

export type ChartCardHeaderMenuAction = ChartCardHeaderActionBase & {
  readonly kind: 'menu';
  readonly icon?: ReactNode;
  readonly items: readonly ChartCardHeaderAction[];
  readonly renderMenu?: (props: {
    readonly action: ChartCardHeaderMenuAction;
    readonly items: readonly ChartCardHeaderAction[];
  }) => ReactNode;
};

export type ChartCardHeaderCustomAction = ChartCardHeaderActionBase & {
  readonly kind: 'custom';
  readonly render: (props?: { readonly closeMenu?: () => void }) => ReactNode;
  readonly renderOverlay?: () => ReactNode;
};

export type ChartCardHeaderAction =
  | ChartCardHeaderButtonAction
  | ChartCardHeaderLinkAction
  | ChartCardHeaderMenuAction
  | ChartCardHeaderCustomAction;

export type ChartCardHeaderActionsProps = {
  readonly actions: readonly ChartCardHeaderAction[];
};

/**
 * Default chart card header action region. Standard actions render with
 * analytics-ui chrome; callers can supply custom renderers for product-owned UI.
 */
const ChartCardHeaderActions: FC<ChartCardHeaderActionsProps> = ({ actions }) => {
  if (!actions.length) {
    return null;
  }

  return (
    <>
      {actions.map((action) => (
        <ChartCardHeaderActionItem action={action} key={action.id} />
      ))}
    </>
  );
};

const ChartCardHeaderActionItem: FC<{ readonly action: ChartCardHeaderAction }> = ({ action }) => {
  switch (action.kind) {
    case 'button':
      return <ChartCardHeaderButton action={action} />;
    case 'link':
      return <ChartCardHeaderLink action={action} />;
    case 'menu':
      return <ChartCardHeaderMenu action={action} />;
    case 'custom':
      return <>{action.render()}</>;
    default:
      return null;
  }
};

const ChartCardHeaderButton: FC<{ readonly action: ChartCardHeaderButtonAction }> = ({
  action,
}) => {
  const { disabled, icon, label, onClick, testId, tooltip } = action;
  const defaultButton = (
    <Tooltip title={disabled ? '' : tooltip} arrow>
      <span>
        <Button
          onClick={onClick}
          variant='contained'
          color='secondary'
          size='small'
          disableRipple
          disabled={disabled}
          className={icon ? 'min-width-fit' : undefined}
          aria-label={icon ? label : undefined}
          data-testid={testId}>
          {icon ?? label}
        </Button>
      </span>
    </Tooltip>
  );

  return <>{action.renderButton?.({ action, defaultButton }) ?? defaultButton}</>;
};

const ChartCardHeaderLink: FC<{ readonly action: ChartCardHeaderLinkAction }> = ({ action }) => {
  const { disabled, href, icon, label, testId, tooltip } = action;
  const defaultLink = (
    <Tooltip title={disabled ? '' : tooltip} arrow>
      <span>
        <Button
          component='a'
          href={disabled ? undefined : href}
          variant='contained'
          color='secondary'
          size='small'
          disableRipple
          disabled={disabled}
          className={icon ? 'min-width-fit' : undefined}
          aria-label={icon ? label : undefined}
          data-testid={testId}>
          {icon ?? label}
        </Button>
      </span>
    </Tooltip>
  );

  return <>{action.renderLink?.({ action, defaultLink }) ?? defaultLink}</>;
};

const ChartCardHeaderMenu: FC<{ readonly action: ChartCardHeaderMenuAction }> = ({ action }) => {
  const { disabled, icon, items, label, renderMenu, testId, tooltip } = action;
  const [anchorEl, setAnchorEl] = useState<HTMLElement | null>(null);
  const isOpen = Boolean(anchorEl);
  const handleOpen = useCallback((event: React.MouseEvent<HTMLButtonElement>) => {
    event.stopPropagation();
    setAnchorEl(event.currentTarget);
  }, []);
  const handleClose = useCallback(() => {
    setAnchorEl(null);
  }, []);

  const defaultMenu = useMemo(
    () => (
      <>
        <Tooltip title={disabled ? '' : tooltip} arrow>
          <span>
            <IconButton
              aria-label={label}
              aria-controls={isOpen ? action.id : undefined}
              aria-haspopup='true'
              aria-expanded={isOpen ? 'true' : undefined}
              color='secondary'
              size='small'
              disabled={disabled}
              onClick={handleOpen}
              data-testid={testId}>
              {icon ?? <MoreVertIcon />}
            </IconButton>
          </span>
        </Tooltip>
        <Menu
          id={action.id}
          anchorEl={anchorEl}
          open={isOpen}
          onClose={handleClose}
          anchorOrigin={{ vertical: 'bottom', horizontal: 'right' }}
          transformOrigin={{ vertical: 'top', horizontal: 'right' }}>
          {items.map((item) => (
            <DefaultMenuItem action={item} key={item.id} onClose={handleClose} />
          ))}
        </Menu>
        {items.map((item) =>
          item.kind === 'custom' && item.renderOverlay ? (
            <React.Fragment key={`${item.id}-overlay`}>{item.renderOverlay()}</React.Fragment>
          ) : null,
        )}
      </>
    ),
    [
      action.id,
      anchorEl,
      disabled,
      handleClose,
      handleOpen,
      icon,
      isOpen,
      items,
      label,
      testId,
      tooltip,
    ],
  );

  return <>{renderMenu?.({ action, items }) ?? defaultMenu}</>;
};

const DefaultMenuItem: FC<{
  readonly action: ChartCardHeaderAction;
  readonly onClose: () => void;
}> = ({ action, onClose }) => {
  const handleButtonClick = useCallback(() => {
    if (action.kind === 'button') {
      action.onClick();
    }
    onClose();
  }, [action, onClose]);

  if (action.kind === 'button') {
    return (
      <MenuItem onClick={handleButtonClick} disabled={action.disabled} data-testid={action.testId}>
        {action.label}
      </MenuItem>
    );
  }

  if (action.kind === 'link') {
    return (
      <MenuItem
        component='a'
        href={action.disabled ? undefined : action.href}
        onClick={onClose}
        disabled={action.disabled}
        data-testid={action.testId}>
        {action.label}
      </MenuItem>
    );
  }

  if (action.kind === 'custom') {
    return <>{action.render({ closeMenu: onClose })}</>;
  }

  return null;
};

export default memo(ChartCardHeaderActions);
