// @rbx/google-gax@0.1.3/build/src/openCloudError.js
"use strict";
/**
 * Copyright 2020 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * Copyright 2023 Roblox Corporation
 *
 * This file was derived and modified from gax/src/googleError.ts
 * in github.com/googleapis/gax-nodejs.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenCloudErrorDecoder = exports.OpenCloudError = void 0;
const status_1 = require("./status");
const protobuf = require("protobufjs");
const serializer = require("proto3-json-serializer");
const fallback_1 = require("./fallback");
class OpenCloudError extends Error {
    // Parse http JSON error and promote google.rpc.ErrorInfo if exist.
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    static parseHttpError(json, status) {
        var _a;
        const decoder = new OpenCloudErrorDecoder();
        const proto3Error = decoder.decodeHTTPError(json);
        const error = Object.assign(new OpenCloudError(json['message']), proto3Error);
        // Infer Status code from HTTP status code if not available in JSON
        (_a = json['code']) !== null && _a !== void 0 ? _a : (json['code'] = status_1.Status[(0, status_1.rpcCodeFromHttpStatusCode)(status)].toString());
        error.code = status_1.Status[json['code']];
        error.message = `${status_1.Status[error.code]}: ${error.message}`;
        error.stack = undefined;
        // Rename "details" to "statusDetails".
        if (error.details) {
            try {
                const statusDetailsObj = decoder.decodeHttpStatusDetails(error.details);
                if (statusDetailsObj &&
                    statusDetailsObj.details &&
                    statusDetailsObj.details.length > 0) {
                    error.statusDetails = statusDetailsObj.details;
                }
                if (statusDetailsObj && statusDetailsObj.errorInfo) {
                    error.reason = statusDetailsObj.errorInfo.reason;
                    error.domain = statusDetailsObj.errorInfo.domain;
                    error.errorInfoMetadata = statusDetailsObj.errorInfo.metadata;
                }
            }
            catch (decodeErr) {
                // ignoring the error
            }
        }
        return error;
    }
}
exports.OpenCloudError = OpenCloudError;
class OpenCloudErrorDecoder {
    constructor() {
        // eslint-disable-next-line @typescript-eslint/no-var-requires
        const errorProtoJson = require('../../build/protos/status.json');
        this.root = protobuf.Root.fromJSON(errorProtoJson);
        this.anyType = this.root.lookupType('google.protobuf.Any');
        this.statusType = this.root.lookupType('google.rpc.Status');
    }
    decodeProtobufAny(anyValue) {
        const match = anyValue.type_url.match(/^type.googleapis.com\/(.*)/);
        if (!match) {
            throw new Error(`Unknown type encoded in google.protobuf.any: ${anyValue.type_url}`);
        }
        const typeName = match[1];
        const type = this.root.lookupType(typeName);
        if (!type) {
            throw new Error(`Cannot lookup type ${typeName}`);
        }
        return type.decode(anyValue.value);
    }
    // Decodes http error which is an instance of google.rpc.Status.
    decodeHTTPError(json) {
        const errorMessage = serializer.fromProto3JSON(this.statusType, json);
        if (!errorMessage) {
            throw new Error(`Received error message ${json}, but failed to serialize as proto3 message`);
        }
        return this.statusType.toObject(errorMessage, fallback_1.defaultToObjectOptions);
    }
    // Decodes http error details which is an instance of Array<google.protobuf.Any>.
    decodeHttpStatusDetails(rawDetails) {
        // eslint-disable-next-line @typescript-eslint/no-unused-vars
        const details = [];
        let errorInfo;
        for (const detail of rawDetails) {
            try {
                const decodedDetail = this.decodeProtobufAny(detail);
                details.push(decodedDetail);
                if (detail.type_url === 'type.googleapis.com/google.rpc.ErrorInfo') {
                    errorInfo = decodedDetail;
                }
            }
            catch (err) {
                // cannot decode detail, likely because of the unknown type - just skip it
            }
        }
        return { details, errorInfo };
    }
}
exports.OpenCloudErrorDecoder = OpenCloudErrorDecoder;
